{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Join Adjacency Matrix, Gene Expression & Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import pandas\n",
    "import numpy as np\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn import preprocessing\n",
    "import matplotlib.pyplot as plt\n",
    "np.set_printoptions(suppress=True)\n",
    "import h5py\n",
    "import mygene\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Get Gene Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# params\n",
    "TIMES_STD_THRESHOLD = 10.\n",
    "MAX_ZEROS_ALLOWED = .5\n",
    "SCALE = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Had gene expression for 41424 genes in the beginning.\n",
      "kicked out 9 super highly expressed genes\n",
      "Kicked out 15021 genes with more than 50.0% zeros\n",
      "==> Left with gene expression for 26394 genes\n"
     ]
    }
   ],
   "source": [
    "# load data\n",
    "gene_expression = pandas.DataFrame.from_csv('../data/preprocessing/normalized-counts-labels.csv',\n",
    "                                            encoding='utf-8', sep=',')\n",
    "ge_nonames = gene_expression.drop('Name', axis=1)\n",
    "\n",
    "# kick out super highly expressed genes\n",
    "threshold = ge_nonames.mean(axis=1).std()*TIMES_STD_THRESHOLD\n",
    "#print (\"Threshold Gene Expression: {}\".format(threshold))\n",
    "anomalies = gene_expression[ge_nonames.mean(axis=1) > threshold]\n",
    "ge_anomalies_removed = ge_nonames.drop(anomalies.index)\n",
    "\n",
    "# kick out genes with too many zeros\n",
    "ge_zeros_removed = ge_anomalies_removed[ge_anomalies_removed.astype('bool').mean(axis=1)>=(1-MAX_ZEROS_ALLOWED)]\n",
    "\n",
    "# write gene expression to file to perform differential expression analysis on it\n",
    "ge_zeros_removed.to_csv('../data/preprocessing/gene_expression_counts.tsv', sep='\\t')\n",
    "\n",
    "# scaling\n",
    "if SCALE:\n",
    "    scaler = preprocessing.StandardScaler()\n",
    "    scaled_features = scaler.fit_transform(ge_zeros_removed)\n",
    "    ge_final = pandas.DataFrame(scaled_features,\n",
    "                                index=ge_zeros_removed.index,\n",
    "                                columns=ge_zeros_removed.columns\n",
    "                               )\n",
    "else:\n",
    "    ge_final = ge_zeros_removed\n",
    "\n",
    "# print some information\n",
    "print (\"Had gene expression for {} genes in the beginning.\".format(ge_nonames.shape[0]))\n",
    "print (\"kicked out {} super highly expressed genes\".format(ge_nonames.shape[0] - ge_anomalies_removed.shape[0]))\n",
    "print (\"Kicked out {} genes with more than {}% zeros\".format(ge_anomalies_removed.shape[0]-ge_zeros_removed.shape[0],\n",
    "                                                             MAX_ZEROS_ALLOWED*100.\n",
    "                                                            ))\n",
    "print (\"==> Left with gene expression for {} genes\".format(ge_final.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Pam3T16</th>\n",
       "      <th>Pam3T8</th>\n",
       "      <th>Pam3T16.1</th>\n",
       "      <th>Pam3T8.1</th>\n",
       "      <th>Pam3T16.2</th>\n",
       "      <th>Pam3T8.2</th>\n",
       "      <th>ControlT8</th>\n",
       "      <th>ControlT16</th>\n",
       "      <th>ControlT8.1</th>\n",
       "      <th>...</th>\n",
       "      <th>gfpmT8.1</th>\n",
       "      <th>gfpmT16.1</th>\n",
       "      <th>gfpmT8.2</th>\n",
       "      <th>gfpmT16.2</th>\n",
       "      <th>gfppT8</th>\n",
       "      <th>gfppT16</th>\n",
       "      <th>gfppT8.1</th>\n",
       "      <th>gfppT16.1</th>\n",
       "      <th>gfppT8.2</th>\n",
       "      <th>gfppT16.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ensembl-ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>6.736327</td>\n",
       "      <td>8.519666</td>\n",
       "      <td>10.773706</td>\n",
       "      <td>10.813725</td>\n",
       "      <td>2.427830</td>\n",
       "      <td>7.425264</td>\n",
       "      <td>9.417385</td>\n",
       "      <td>7.920964</td>\n",
       "      <td>10.812953</td>\n",
       "      <td>...</td>\n",
       "      <td>12.529836</td>\n",
       "      <td>6.769370</td>\n",
       "      <td>13.440488</td>\n",
       "      <td>5.248289</td>\n",
       "      <td>14.583375</td>\n",
       "      <td>13.538040</td>\n",
       "      <td>18.220493</td>\n",
       "      <td>14.908853</td>\n",
       "      <td>16.156797</td>\n",
       "      <td>30.745221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>97.676739</td>\n",
       "      <td>73.025712</td>\n",
       "      <td>132.362676</td>\n",
       "      <td>110.840686</td>\n",
       "      <td>87.401897</td>\n",
       "      <td>87.865624</td>\n",
       "      <td>92.725018</td>\n",
       "      <td>105.132797</td>\n",
       "      <td>121.218890</td>\n",
       "      <td>...</td>\n",
       "      <td>127.983328</td>\n",
       "      <td>147.421826</td>\n",
       "      <td>104.163784</td>\n",
       "      <td>86.596770</td>\n",
       "      <td>298.959192</td>\n",
       "      <td>239.172041</td>\n",
       "      <td>268.101547</td>\n",
       "      <td>155.611158</td>\n",
       "      <td>283.898001</td>\n",
       "      <td>247.298513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>84.204085</td>\n",
       "      <td>74.242807</td>\n",
       "      <td>61.564035</td>\n",
       "      <td>68.486928</td>\n",
       "      <td>78.499852</td>\n",
       "      <td>92.815800</td>\n",
       "      <td>86.929705</td>\n",
       "      <td>72.728852</td>\n",
       "      <td>66.585024</td>\n",
       "      <td>...</td>\n",
       "      <td>76.074006</td>\n",
       "      <td>63.180782</td>\n",
       "      <td>117.604272</td>\n",
       "      <td>94.469203</td>\n",
       "      <td>83.125239</td>\n",
       "      <td>79.724014</td>\n",
       "      <td>76.786365</td>\n",
       "      <td>62.430824</td>\n",
       "      <td>85.400212</td>\n",
       "      <td>93.572410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>61.468982</td>\n",
       "      <td>48.683808</td>\n",
       "      <td>70.798641</td>\n",
       "      <td>63.080065</td>\n",
       "      <td>63.123592</td>\n",
       "      <td>76.727728</td>\n",
       "      <td>68.819350</td>\n",
       "      <td>68.408327</td>\n",
       "      <td>77.397977</td>\n",
       "      <td>...</td>\n",
       "      <td>56.384263</td>\n",
       "      <td>72.958761</td>\n",
       "      <td>81.482960</td>\n",
       "      <td>51.608176</td>\n",
       "      <td>24.791738</td>\n",
       "      <td>33.092987</td>\n",
       "      <td>29.933668</td>\n",
       "      <td>59.635414</td>\n",
       "      <td>30.005480</td>\n",
       "      <td>26.734974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>FGR</td>\n",
       "      <td>53.890614</td>\n",
       "      <td>150.919804</td>\n",
       "      <td>143.136382</td>\n",
       "      <td>362.259802</td>\n",
       "      <td>156.999704</td>\n",
       "      <td>160.880721</td>\n",
       "      <td>62.299622</td>\n",
       "      <td>61.207450</td>\n",
       "      <td>189.511222</td>\n",
       "      <td>...</td>\n",
       "      <td>333.830638</td>\n",
       "      <td>257.988195</td>\n",
       "      <td>186.486774</td>\n",
       "      <td>70.851902</td>\n",
       "      <td>129.792040</td>\n",
       "      <td>97.774734</td>\n",
       "      <td>272.005939</td>\n",
       "      <td>245.996082</td>\n",
       "      <td>140.794944</td>\n",
       "      <td>101.592903</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Name    Pam3T16      Pam3T8   Pam3T16.1    Pam3T8.1  \\\n",
       "Ensembl-ID                                                                 \n",
       "ENSG00000000003    TSPAN6   6.736327    8.519666   10.773706   10.813725   \n",
       "ENSG00000000419      DPM1  97.676739   73.025712  132.362676  110.840686   \n",
       "ENSG00000000457     SCYL3  84.204085   74.242807   61.564035   68.486928   \n",
       "ENSG00000000460  C1orf112  61.468982   48.683808   70.798641   63.080065   \n",
       "ENSG00000000938       FGR  53.890614  150.919804  143.136382  362.259802   \n",
       "\n",
       "                  Pam3T16.2    Pam3T8.2  ControlT8  ControlT16  ControlT8.1  \\\n",
       "Ensembl-ID                                                                    \n",
       "ENSG00000000003    2.427830    7.425264   9.417385    7.920964    10.812953   \n",
       "ENSG00000000419   87.401897   87.865624  92.725018  105.132797   121.218890   \n",
       "ENSG00000000457   78.499852   92.815800  86.929705   72.728852    66.585024   \n",
       "ENSG00000000460   63.123592   76.727728  68.819350   68.408327    77.397977   \n",
       "ENSG00000000938  156.999704  160.880721  62.299622   61.207450   189.511222   \n",
       "\n",
       "                    ...        gfpmT8.1   gfpmT16.1    gfpmT8.2  gfpmT16.2  \\\n",
       "Ensembl-ID          ...                                                      \n",
       "ENSG00000000003     ...       12.529836    6.769370   13.440488   5.248289   \n",
       "ENSG00000000419     ...      127.983328  147.421826  104.163784  86.596770   \n",
       "ENSG00000000457     ...       76.074006   63.180782  117.604272  94.469203   \n",
       "ENSG00000000460     ...       56.384263   72.958761   81.482960  51.608176   \n",
       "ENSG00000000938     ...      333.830638  257.988195  186.486774  70.851902   \n",
       "\n",
       "                     gfppT8     gfppT16    gfppT8.1   gfppT16.1    gfppT8.2  \\\n",
       "Ensembl-ID                                                                    \n",
       "ENSG00000000003   14.583375   13.538040   18.220493   14.908853   16.156797   \n",
       "ENSG00000000419  298.959192  239.172041  268.101547  155.611158  283.898001   \n",
       "ENSG00000000457   83.125239   79.724014   76.786365   62.430824   85.400212   \n",
       "ENSG00000000460   24.791738   33.092987   29.933668   59.635414   30.005480   \n",
       "ENSG00000000938  129.792040   97.774734  272.005939  245.996082  140.794944   \n",
       "\n",
       "                  gfppT16.2  \n",
       "Ensembl-ID                   \n",
       "ENSG00000000003   30.745221  \n",
       "ENSG00000000419  247.298513  \n",
       "ENSG00000000457   93.572410  \n",
       "ENSG00000000460   26.734974  \n",
       "ENSG00000000938  101.592903  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_names = gene_expression.loc[ge_final.index].Name.values\n",
    "gene_ids = ge_final.index\n",
    "gene_expression.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## ConsensusPathDB: Get PPI Network\n",
    "I want to construct an adjacency matrix from the interactions in ConsensusPathDB.\n",
    "However, I only want it for the genes for which I have gene expression data. If not both genes are involved in an interaction, I don't want to save the interaction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We lose 57141 interactions by removing those where not both genes have expression\n",
      "Still keeping 161231 interactions\n"
     ]
    }
   ],
   "source": [
    "# read interaction data (implicit threshold of 0.5 for interaction confidence)\n",
    "\n",
    "interactions = pandas.read_csv('../data/networks/CPDB_ensg_edgelist.tsv',\n",
    "                               header=0,\n",
    "                               sep='\\t',\n",
    "                               names=['number', 'gene1', 'gene2', 'score']\n",
    "                              )\n",
    "\n",
    "# kick out interactions for which we don't have gene expression\n",
    "lost = interactions[~interactions.gene1.isin(ge_final.index) | ~interactions.gene2.isin(ge_final.index)].shape[0]\n",
    "print (\"We lose {} interactions by removing those where not both genes have expression\".format(lost))\n",
    "interactions_ge = interactions[interactions.gene1.isin(ge_final.index) & interactions.gene2.isin(ge_final.index)]\n",
    "print (\"Still keeping {} interactions\".format(interactions_ge.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed Adjacency Matrix with average node degree of: 15.505962685131756\n",
      "Adjacency matrix has 161231.0 edges in total\n",
      "Network has 10398 nodes in total\n"
     ]
    }
   ],
   "source": [
    "# construct adjacency matrix (memory consuming)\n",
    "ppi_network = interactions_ge.pivot('gene1', 'gene2', 'score')\n",
    "ppi_network[~ppi_network.isnull()] = 1\n",
    "ppi_network[ppi_network.isnull()] = 0\n",
    "\n",
    "# add genes for which no interactions are present\n",
    "index = ppi_network.index.union(ppi_network.columns)\n",
    "ppi_network = ppi_network.reindex(index=index, columns=index, fill_value=0)\n",
    "\n",
    "# output some statistics\n",
    "avg_node_degree = ppi_network.sum().mean()\n",
    "total_num_of_connections = ppi_network.sum().sum()\n",
    "\n",
    "print (\"Constructed Adjacency Matrix with average node degree of: {}\".format(avg_node_degree))\n",
    "print (\"Adjacency matrix has {} edges in total\".format(total_num_of_connections))\n",
    "print (\"Network has {} nodes in total\".format(ppi_network.index.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x7fe380af47b8>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA00AAAHwCAYAAAB6yISuAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xu4dWVZL/7vLXhEEhQ05SCYlFF52oiYZop5PuA2Nc2S\niMT2j59h6VY0C4+ltT2muUXR8Cx5AhVTRPFQqYCoiGaiYYAoKIioiQL3/mOOpbPXtcaaL75zrbne\n9/O5rnWtMZ5xuucak8X6vs8znlndHQAAAJZ3jfUuAAAAYJEJTQAAACOEJgAAgBFCEwAAwAihCQAA\nYITQBAAAMEJoAtiCqurpVfX69a5jUVXVe6vq4C10rt+oqi9OrZ9TVb+1Jc49nO+sqrrbljrfjNes\nqnpNVV1SVZ9cy2tvSVX11Kp61XrXAbClCE3ANm34Q/vCqtphqu2PquqUdarlv6rqsqr6dlX9S1X9\ncVVtiN/VVdVV9b2q+m5VfauqTq6q35nep7vv293HzniuW47t090f7e5f+lnrHq73D1X17E3O/yvd\nfcqWOP9muEuSeybZvbv3X26HqrppVb2yqr42/Ky/MtR/q7UtdWXd/Vfd/UdX59jhtfxweG0XV9VJ\nS69t+EeJHw3blv4budOw7Q+q6mNb8nUALNkQ/yMGmLPtkhyx3kUMHtjdOya5eZLnJnlykmPmcaGq\n2m4Op71Nd18/yS8l+YckL62qo7b0Rapq+y19zgVx8yTndPf3lttYVTdK8i9JrpfkN5LsmOT2ST6c\nSdjaWvzN8D7aPcmFmbyXlrxl2LZrko8leXtV1dqXCGxLhCaA5G+TPLGqdlpuY1X9elWdWlWXDt9/\nfWrb3lX14aF36KQku2xy7AHDv4Z/u6o+M+twr+6+tLtPSPI7SQ6uql8dznftqvo/VfWfVfWNqvq/\nVXXdqes9qaouGHoh/mi6x2b4F/yXV9WJVfW9JHef4XwPqKpPT/2r/q1nrP+b3f26JP8ryVOGP/ZT\nVadU1R8Ny7ccfnaXVtU3q+otQ/tHhtN8ZuhR+J2qultVnVdVT66qryd5zVLbJpe+Q1V9fhje9pqq\nus5wzp/qhVj62VTVYUkeleRJw/XeNWz/8XC/4ef0ouHn+rVh+drDtqXanlCTXssLquqQlX42VXWz\nqjph6EU5u6oeM7QfmuRVSe401PGMZQ7/0yTfSfL73f3lnvh2d7+mu/9u6horvu+Ge/Csqvrn4X37\n/qraZcZj/6AmPVuXVdV/VNWjVniNPx6mWlV7DT/rg4f32Ter6s9X+vlM6+7vJ3ljkl9dZtuPkhyb\n5OeT3GiW8wFcXUITQHJaklOSPHHTDVV1wyTvSfKSTP4we0GS9yyFgEz+oDs9k7D0rCQHTx2723Ds\ns5PccDj/26pq11kL6+5PJjkvk16FZNL79ItJbpvklkl2S/KXw/Xuk+TPkvzWsO1uy5zyd5M8J5Me\nio+tcr7bJXl1kscOr/0VSU5YCgszOj7J9kmWG2r2rCTvT7JzJj0Kfze85rsO22/T3dfv7rcM6z+f\nyc/x5kkOW+F6j0py7yS/MLyup61WYHcfneQNGXo3uvuBy+z250kOyOTndJvh9Uyf++eT3CCTn9+h\nSV5WVTuvcMk3Z3JPb5bkoUn+qqoO7O5jkvxxkn8d6liuh+63kryju69a6fXM+L773SSHJLlxkmsN\n+4weW5MhrC9Jct+hN/TXk3x6pTqWcZdMeiDvkeQvq+qXVzugqq6fyT09Y5lt107yB0nO7e5vbkYd\nAJtNaAKY+Mskj1sm0Nw/yZe6+3XdfUV3vynJvyV5YFXtmeQOSf6iuy/v7o8kedfUsb+X5MTuPrG7\nr+rukzIJaPfbzNq+luSGVVWZhIU/7e6Lu/uyJH+V5BHDfg9P8pruPmv4F/qnL3Ou47v7n4c/ui9f\n5XyHJXlFd3+iu68cnkW6PJPwMJOhN+CbmfwBvqkfZRKAbtbdP+ju1Z5HuSrJUcPP+r9W2Oel3X1u\nd1+cSTh85Ky1ruJRSZ7Z3Rd290VJnpHk96e2/2jY/qPuPjHJdzMJCP9NVe2R5M5Jnjy85k9n0rv0\n6Bnr2CXJ16fO96ChR+iyqnr/0DzL++413f3vw8/xuEzC4CzHXpXkV6vqut19QXefNWPdSfKM7v6v\n7v5Mks9kEj5X8sSq+naSs5NcP5NwtOThw7Zzk/yPJP9zM2oAuFqEJoAk3f25JO9OcuQmm26W5Kub\ntH01kx6FmyW5ZJPnT6b3vXmShw1/1H57+EPvLkluupnl7Zbk4kye4bhektOnzvdPQ/tSredOHXdu\nftp022rnu3mSJ2xS/x7DdWZSVdccznfxMpuflKSSfLImM9X94Sqnu6i7f7DKPtOv76ubU+sqNn0f\nbHrub3X3FVPr38/kj/3lzrMUUKfPtduMdXwrU++f7j6hu3fKZNjetYbmWd53X59anq51xWOH9/nv\nZNIbdkFVvac2b/KJla65nP/T3Tt1989394O6+8tT244btt24uw/s7tM3owaAq2VrfZAW4Oo4Ksmn\nkjx/qu1rmfwhOW3PTMLFBUl2rqodpoLTnkl6WD43yeu6+zFXt6CqukMmf1B/LJMem/9K8ivdff4y\nu1+QyTC3JXsss09PLa92vnOTPKe7n3N1ah8clOSKJD81fXZ3fz3J0vM8d0nygar6SHefvcK5eoX2\nadOvec9M7l+SfC+TgJjhej+/medeeh8s9axMn3tzLPUa7jgVnPZMstzPfzknJ3lwVT1jZIjez/K+\nGz22u9+X5H01ee7t2UlemZ8MHQXYaulpAhgMf6y/JcmfTDWfmOQXq+p3q2r7mkyhvW+Sd3f3VzMZ\nuvSMqrrW8If/9PMwr89kGN+9q2q7qrrOMGnAdLBZVlX9XFU9IJPnX17f3WcOfyS/MskLq+rGw367\nVdW9h8OOS3JIVf1yVV0vyV+s8npXO98rk/xxVd2xJnaoqvtX1Y4z1H/DYZKAlyV5Xnd/a5l9Hjb1\ns7gkk+CyFAS+keQWq11nGYdX1e7Ds2h/nsn9TCbDwX6lqm5bk8khnr7Jcatd701JnjY827NLJsM5\nN/vzuLr73Exmv/vr4f1w60yegZr1XC/I5Bmw11XVLwz3Zcf8ZHhd8jO878aOraqbVNVBw7NNl2cy\nBHHFZ6sAtiZCE8B/98wkP/7MpuGP/QckeUImQ6OelOQBUw+e/26SO2Yy/OyoJK+dOvbcTHpanprk\nokz+Ff9/Z/x377uq6rJh3z/P5I/k6ZnYnpzJcx4fr6rvJPlAhmdnuvu9mTyo/6GlfYZjLh+53tj5\nTsukJ+ilmYSas/Pfny1Zzmeq6rvDvn+UyfNSf7nCvndI8olh/xOSHNHdXxm2PT3JscMQsYevcs1p\nb8xkcomvJPlyJr0h6e5/z+TefiDJlzLpuZt2TJJ9h+u9c5nzPjuTgPzZJGdm0iP57GX2m8Ujk+yV\nSa/TOzJ5TusDsxw4vO8OSPKDTF7DZZlMxrBjJjMVXt333dL5x469RiYTjXwtk/f7by5dE2BrV92z\njHYAYKMZZif7XJJrb/K8DQCwGfQ0AWxFqup/1uQzhXZO8rwk7xKYAOBnIzQBbF0em+TCTIamXRnD\npwDgZ2Z4HgAAwAg9TQAAACOEJgAAgBFz/XDbqjonk+lQr0xyRXfvN3x2xlsymW71nCQP7+5LqqqS\nvDjJ/TL5pPA/6O5PDec5OMnThtM+u7uPHbvuLrvs0nvttdcWfz0AAMDW4/TTT/9md++62n5zDU2D\nu099nkmSHJnk5O5+blUdOaw/Ocl9k+wzfN0xycuT3HEIWUcl2S+TDz48vapO6O5LVrrgXnvtldNO\nO20+rwYAANgqVNVXZ9lvPYbnHZRkqafo2CQPnmp/bU98PMlOVXXTJPdOclJ3XzwEpZOS3GetiwYA\nALZN8w5NneT9VXV6VR02tN2kuy8Ylr+e5CbD8m6ZfPL4kvOGtpXaAQAA5m7ew/Pu0t3nV9WNk5xU\nVf82vbG7u6q2yJznQyg7LEn23HPPLXFKAACA+fY0dff5w/cLk7wjyf5JvjEMu8vw/cJh9/OT7DF1\n+O5D20rtm17r6O7er7v323XXVZ/lAgAAmMncQlNV7VBVOy4tJ7lXks8lOSHJwcNuByc5flg+Icmj\na+KAJJcOw/jel+ReVbVzVe08nOd986obAABg2jyH590kyTsmM4ln+yRv7O5/qqpTkxxXVYcm+WqS\nhw/7n5jJdONnZzLl+CFJ0t0XV9Wzkpw67PfM7r54jnUDAAD8WHVvkUeKFsp+++3XphwHAADGVNXp\n3b3favutx5TjAAAAG4bQBAAAMEJoAgAAGCE0AQAAjBCaAAAARghNAAAAI4QmAACAEUITAADACKEJ\nAABghNAEAAAwQmgCAAAYITQBAACM2H69C9gW7HXke1bcds5z77+GlQAAAJtLTxMAAMAIoQkAAGCE\n0AQAADBCaAIAABghNAEAAIwQmgAAAEYITQAAACOEJgAAgBFCEwAAwAihCQAAYITQBAAAMEJoAgAA\nGCE0AQAAjBCaAAAARghNAAAAI4QmAACAEUITAADACKEJAABghNAEAAAwQmgCAAAYITQBAACMEJoA\nAABGCE0AAAAjhCYAAIARQhMAAMAIoQkAAGCE0AQAADBCaAIAABghNAEAAIwQmgAAAEYITQAAACOE\nJgAAgBFCEwAAwAihCQAAYITQBAAAMEJoAgAAGCE0AQAAjBCaAAAARghNAAAAI4QmAACAEUITAADA\nCKEJAABghNAEAAAwQmgCAAAYITQBAACMEJoAAABGCE0AAAAjhCYAAIARQhMAAMAIoQkAAGCE0AQA\nADBCaAIAABghNAEAAIwQmgAAAEYITQAAACOEJgAAgBFCEwAAwAihCQAAYITQBAAAMEJoAgAAGCE0\nAQAAjBCaAAAARghNAAAAI4QmAACAEXMPTVW1XVWdUVXvHtb3rqpPVNXZVfWWqrrW0H7tYf3sYfte\nU+d4ytD+xaq697xrBgAAWLIWPU1HJPnC1Przkrywu2+Z5JIkhw7thya5ZGh/4bBfqmrfJI9I8itJ\n7pPk76tquzWoGwAAYL6hqap2T3L/JK8a1ivJgUneOuxybJIHD8sHDesZtt9j2P+gJG/u7su7+z+S\nnJ1k/3nWDQAAsGTePU0vSvKkJFcN6zdK8u3uvmJYPy/JbsPybknOTZJh+6XD/j9uX+YYAACAuZpb\naKqqByS5sLtPn9c1NrneYVV1WlWddtFFF63FJQEAgG3APHua7pzkQVV1TpI3ZzIs78VJdqqq7Yd9\ndk9y/rB8fpI9kmTYfoMk35puX+aYH+vuo7t7v+7eb9ddd93yrwYAANgmzS00dfdTunv37t4rk4kc\nPtjdj0ryoSQPHXY7OMnxw/IJw3qG7R/s7h7aHzHMrrd3kn2SfHJedQMAAEzbfvVdtrgnJ3lzVT07\nyRlJjhnaj0nyuqo6O8nFmQStdPdZVXVcks8nuSLJ4d195dqXDQAAbIvWJDR19ylJThmWv5JlZr/r\n7h8kedgKxz8nyXPmVyEAAMDy1uJzmgAAADYsoQkAAGCE0AQAADBCaAIAABghNAEAAIwQmgAAAEYI\nTQAAACOEJgAAgBFCEwAAwAihCQAAYITQBAAAMEJoAgAAGCE0AQAAjBCaAAAARghNAAAAI4QmAACA\nEUITAADACKEJAABghNAEAAAwQmgCAAAYITQBAACMEJoAAABGCE0AAAAjhCYAAIARQhMAAMAIoQkA\nAGCE0AQAADBCaAIAABghNAEAAIwQmgAAAEYITQAAACOEJgAAgBFCEwAAwAihCQAAYITQBAAAMGKm\n0FRVN6+q3xqWr1tVO863LAAAgMWwamiqqsckeWuSVwxNuyd55zyLAgAAWBSz9DQdnuTOSb6TJN39\npSQ3nmdRAAAAi2KW0HR5d/9waaWqtk/S8ysJAABgccwSmj5cVU9Nct2qumeSf0zyrvmWBQAAsBhm\nCU1HJrkoyZlJHpvkxCRPm2dRAAAAi2L71Xbo7quq6vVJPtLdX1yDmgAAABbGLLPnPSjJp5P807B+\n26o6Yd6FAQAALIJZhucdlWT/JN9Oku7+dJK951kUAADAopglNP2ouy/dpM3seQAAwDZh1WeakpxV\nVb+bZLuq2ifJnyT5l/mWBQAAsBhm6Wl6XJJfSXJ5kjcmuTTJ4+dZFAAAwKIY7Wmqqu2SPLO7n5jk\nz9emJAAAgMUx2tPU3Vcmucsa1QIAALBwZnmm6YxhivF/TPK9pcbufvvcqgIAAFgQs4Sm6yT5VpID\np9o6idAEAABs9VYNTd19yFoUAgAAsIhWDU1V9ZJlmi9Nclp3H7/lSwIAAFgcs0w5fp0kt03ypeHr\n1kl2T3JoVb1ojrUBAACsu1meabp1kjsPM+mlql6e5KOZzKp35hxrAwAAWHez9DTtnOT6U+s7JLnh\nEKIun0tVAAAAC2KWnqa/SfLpqjolSSW5a5K/qqodknxgjrUBAACsu1lmzzumqk5Msv/Q9NTu/tqw\n/L/nVhkAAMACWHV4XlVVknskuc0wW972VbX/KocBAABsFWZ5punvk9wpySOH9cuSvGxuFQEAACyQ\nWZ5pumN3376qzkiS7r6kqq4157oAAAAWwiw9TT+qqu2SdJJU1a5JrpprVQAAAAtiltD0kiTvSHLj\nqnpOko8l+au5VgUAALAgZpk97w1VdXomk0FUkgd39xfmXhkAAMACWDE0VdUNp1YvTPKm6W3dffE8\nCwMAAFgEYz1Np2fyHFMl2TPJJcPyTkn+M8nec68OAABgna34TFN3793dt0jygSQP7O5duvtGSR6Q\n5P1rVSAAAMB6mmUiiAO6+8Slle5+b5Jfn19JAAAAi2OWz2n6WlU9Lcnrh/VHJfna/EoCAABYHLP0\nND0yya6ZTDv+9mH5kfMsCgAAYFHMMuX4xUmOWINaAAAAFs4sPU0AAADbLKEJAABghNAEAAAwYtXQ\nVFW/WFUnV9XnhvVbD7PpAQAAbPVm6Wl6ZZKnJPlRknT3Z5M8Yp5FAQAALIpZQtP1uvuTm7RdMY9i\nAAAAFs0soembVfULSTpJquqhSS5Y7aCquk5VfbKqPlNVZ1XVM4b2vavqE1V1dlW9paquNbRfe1g/\ne9i+19S5njK0f7Gq7n01XicAAMDVMktoOjzJK5LcqqrOT/L4JP9rhuMuT3Jgd98myW2T3KeqDkjy\nvCQv7O5bJrkkyaHD/ocmuWRof+GwX6pq30yGA/5Kkvsk+fuq2m7G1wcAAPAzWTU0dfdXuvu3kuya\n5FbdfZfuPmeG47q7vzusXnP46iQHJnnr0H5skgcPywcN6xm236Oqamh/c3df3t3/keTsJPvP8uIA\nAAB+VtuvtKGq/myF9iRJd79gtZMPPUKnJ7llkpcl+XKSb3f30jNR5yXZbVjeLcm5w7mvqKpLk9xo\naP/41GmnjwEAAJirFUNTkh2H77+U5A5JThjWH5hk04khltXdVya5bVXtlOQdSW51NetcVVUdluSw\nJNlzzz3ndRkAAGAbs2Jo6u6liRs+kuT23X3ZsP70JO/ZnIt097er6kNJ7pRkp6rafuht2j3J+cNu\n5yfZI8l5VbV9khsk+dZU+5LpY6avcXSSo5Nkv/32682pDwAAYCWzTARxkyQ/nFr/4dA2qqp2HXqY\nUlXXTXLPJF9I8qEkDx12OzjJ8cPyCcN6hu0f7O4e2h8xzK63d5J9MmNPFwAAwM9qbHjektcm+WRV\nvSPJ0sQM/zDDcTdNcuzwXNM1khzX3e+uqs8neXNVPTvJGUmOGfY/JsnrqursJBdn+ADd7j6rqo5L\n8vlMPh/q8GHYHwAAwNytGpq6+zlV9d4kv5HJ7HeHdPcZMxz32SS3W6b9K1lm9rvu/kGSh61UQ5Ln\nrHZNAACALW2WnqYkuTLJVZmEpqvmVw4AAMBiWfWZpqo6IskbkuyS5MZJXl9Vj5t3YQAAAItglp6m\nQ5Pcsbu/lyRV9bwk/5rk7+ZZGAAAwCKYZfa8ymR43pIrhzYAAICt3iw9Ta9J8olh9rwkeXB+MuMd\nAADAVm2W2fNeUFUfTnLnoWmm2fMAAAC2BrPOnvfpJBcs7V9Ve3b3f86tKgAAgAWxamgaZso7Ksk3\n8pPnmTrJredbGgAAwPqbpafpiCS/1N3fmncxAAAAi2aW2fPOTXLpvAsBAABYRLP0NH0lySlV9Z4k\nly81dvcL5lYVAADAgpglNP3n8HWt4QsAAGCbMcuU489Yi0IAAAAW0SzPNAEAAGyzhCYAAIARQhMA\nAMCIWT7cdtckj0my1/T+3f2H8ysLAABgMcwye97xST6a5ANJrpxvOQAAAItlltB0ve5+8twrAQAA\nWECzPNP07qq639wrAQAAWEAr9jRV1WVJOkkleWpVXZ7kR8N6d/fPrU2JAAAA62fF0NTdO65lIQAA\nAIto1eF5VXXyLG0AAABbo7HheddJskOSXapq50yG5SXJzyXZbQ1qAwAAWHdjs+c9Nsnjk9wsyaem\n2r+T5KXzLAoAAGBRjD3T9OIkL66qx3X3361hTQAAAAtjbHjegd39wSTnV9VDNt3e3W+fa2UAAAAL\nYGx43m8m+WCSBy6zrZMITQAAwFZvbHjeUcP3Q9auHAAAgMUy1tOUJKmqLyf5eJKPJvlod58196oA\nAAAWxKqf05Rk3ySvSHKjJH9bVV+uqnfMtywAAIDFMEtoujLJj4bvVyW5cPgCAADY6q06PC+Tz2U6\nM8kLkryyu78135IAAAAWxyw9TY9M8pEk/1+SN1fVM6rqHvMtCwAAYDGs2tPU3ccnOb6qbpXkvkke\nn+RJSa4759oAAADW3ao9TVX1tqo6O8mLk1wvyaOT7DzvwgAAABbBLM80/XWSM7r7ynkXAwAAsGhm\nGZ532loUAgAAsIhmmQgCAABgmyU0AQAAjJhlIog7V9UOw/LvVdULqurm8y8NAABg/c3S0/TyJN+v\nqtskeUKSLyd57VyrAgAAWBCzhKYruruTHJTkpd39siQ7zrcsAACAxTDLlOOXVdVTkvx+kt+oqmsk\nueZ8ywIAAFgMs/Q0/U6Sy5P8YXd/PcnuSf52rlUBAAAsiFVD0xCU3pbk2kPTN5O8Y55FAQAALIpZ\nZs97TJK3JnnF0LRbknfOsygAAIBFMcvwvMOT3DnJd5Kku7+U5MbzLAoAAGBRzBKaLu/uHy6tVNX2\nSXp+JQEAACyOWULTh6vqqUmuW1X3TPKPSd4137IAAAAWwyyh6cgkFyU5M8ljk5yY5GnzLAoAAGBR\nrPo5Td19VZJXDl8AAADblFVDU1XdOcnTk9x82L+SdHffYr6lAQAArL9VQ1OSY5L8aZLTk1w533IA\nAAAWyyyh6dLufu/cKwEAAFhAs4SmD1XV3yZ5e5LLlxq7+1NzqwoAAGBBzBKa7jh832+qrZMcuOXL\nAQAAWCyzzJ5397UoBAAAYBGt+jlNVXWTqjqmqt47rO9bVYfOvzQAAID1N8uH2/5Dkvcludmw/u9J\nHj+vggAAABbJLKFpl+4+LslVSdLdV8TU4wAAwDZiltD0vaq6USaTP6SqDkhy6VyrAgAAWBCzzJ73\nZ0lOSPILVfXPSXZN8tC5VgUAALAgZpk971NV9ZtJfilJJflid/9o7pUBAAAsgFVDU1U9ZJOmX6yq\nS5Oc2d0XzqcsAACAxTDL8LxDk9wpyYeG9bslOT3J3lX1zO5+3ZxqAwAAWHezhKbtk/xyd38jmXxu\nU5LXJrljko8kEZoAAICt1iyz5+2xFJgGFw5tFyfxbBMAALBVm6Wn6ZSqeneSfxzWf3to2yHJt+dW\nGQAAwAKYJTQdnuQhSe4yrL82ydu6u5PcfV6FAQAALIJZphzvJG8bvgAAALYpszzTBAAAsM0SmgAA\nAEasGJqq6uTh+/PWrhwAAIDFMvZM002r6teTPKiq3pykpjd296fmWhkAAMACGAtNf5nkL5LsnuQF\nm2zrJAfOqygAAIBFsWJo6u63JnlrVf1Fdz9rDWsCAABYGLNMOf6sqnpQkrsOTad097vnWxYAAMBi\nWHX2vKr66yRHJPn88HVEVf3VDMftUVUfqqrPV9VZVXXE0H7Dqjqpqr40fN95aK+qeklVnV1Vn62q\n20+d6+Bh/y9V1cFX98UCAABsrlmmHL9/knt296u7+9VJ7pPkATMcd0WSJ3T3vkkOSHJ4Ve2b5Mgk\nJ3f3PklOHtaT5L5J9hm+Dkvy8mQSspIcleSOSfZPctRS0AIAAJi3WT+naaep5RvMckB3X7A0w153\nX5bkC0l2S3JQkmOH3Y5N8uBh+aAkr+2JjyfZqapumuTeSU7q7ou7+5IkJ2US3AAAAOZu1Weakvx1\nkjOq6kOZTDt+1/ykd2gmVbVXktsl+USSm3T3BcOmrye5ybC8W5Jzpw47b2hbqX3TaxyWSQ9V9txz\nz80pDwAAYEWzTATxpqo6JckdhqYnd/fXZ71AVV0/yduSPL67v1P1k4976u6uqt68kles8+gkRyfJ\nfvvtt0XOCQAAMEtPU4aeoRM29+RVdc1MAtMbuvvtQ/M3quqm3X3BMPzuwqH9/CR7TB2++9B2fpK7\nbdJ+yubWAgAAcHXM+kzTZqtJl9IxSb7Q3dMfjntCkqUZ8A5OcvxU+6OHWfQOSHLpENbel+ReVbXz\nMAHEvYY2AACAuZupp+lqunOS309yZlV9emh7apLnJjmuqg5N8tUkDx+2nZjkfknOTvL9JIckSXdf\nXFXPSnLqsN8zu/viOdYNAADwY6Ohqaq2S3JWd99qc0/c3R/LZOKI5dxjmf07yeErnOvVSV69uTUA\nAAD8rEaH53X3lUm+WFWmowMAALZJswzP2znJWVX1ySTfW2rs7gfNrSoAAIAFMUto+ou5VwEAALCg\nZvmcpg9X1c2T7NPdH6iq6yXZbv6lAQAArL9VpxyvqsckeWuSVwxNuyV55zyLAgAAWBSzfE7T4ZlM\nH/6dJOnuLyW58TyLAgAAWBSzhKbLu/uHSytVtX2Snl9JAAAAi2OW0PThqnpqkutW1T2T/GOSd823\nLAAAgMUwS2g6MslFSc5M8tgkJyZ52jyLAgAAWBSzzJ53VVUdm+QTmQzL+2J3G54HAABsE1YNTVV1\n/yT/N8mVllEmAAAP6ElEQVSXk1SSvavqsd393nkXBwAAsN5m+XDb5ye5e3efnSRV9QtJ3pNEaAIA\nALZ6szzTdNlSYBp8Jcllc6oHAABgoazY01RVDxkWT6uqE5Mcl8kzTQ9Lcuoa1AYAALDuxobnPXBq\n+RtJfnNYvijJdedWEQAAwAJZMTR19yFrWQgAAMAimmX2vL2TPC7JXtP7d/eD5lcWAADAYphl9rx3\nJjkmybuSXDXfcgAAABbLLKHpB939krlXAgAAsIBmCU0vrqqjkrw/yeVLjd39qblVBQAAsCBmCU2/\nluT3kxyYnwzP62EdAABgqzZLaHpYklt09w/nXQwAAMCiucYM+3wuyU7zLgQAAGARzdLTtFOSf6uq\nU/Pfn2ky5TgAALDVmyU0HTX3KgAAABbUqqGpuz+8FoUAAAAsolVDU1VdlslseUlyrSTXTPK97v65\neRYGAACwCGbpadpxabmqKslBSQ6YZ1EAAACLYpbZ836sJ96Z5N5zqgcAAGChzDI87yFTq9dIsl+S\nH8ytIgAAgAUyy+x5D5xaviLJOZkM0QMAANjqzfJM0yFrUQgAAMAiWjE0VdVfjhzX3f2sOdQDAACw\nUMZ6mr63TNsOSQ5NcqMkQhMAALDVWzE0dffzl5arasckRyQ5JMmbkzx/peMAAAC2JqPPNFXVDZP8\nWZJHJTk2ye27+5K1KAwAAGARjD3T9LdJHpLk6CS/1t3fXbOqtiF7HfmeFbed89z7r2ElAADAcsY+\n3PYJSW6W5GlJvlZV3xm+Lquq76xNeQAAAOtr7JmmsUAFAACwTRCMAAAARghNAAAAI4QmAACAEUIT\nAADACKEJAABghNAEAAAwQmgCAAAYITQBAACMEJoAAABGCE0AAAAjhCYAAIARQhMAAMAIoQkAAGCE\n0AQAADBCaAIAABghNAEAAIwQmgAAAEYITQAAACOEJgAAgBFCEwAAwAihCQAAYITQBAAAMEJoAgAA\nGCE0AQAAjBCaAAAARghNAAAAI4QmAACAEUITAADACKEJAABghNAEAAAwQmgCAAAYITQBAACMEJoA\nAABGCE0AAAAjhCYAAIARQhMAAMAIoQkAAGDE9utdACvb68j3rLjtnOfefw0rAQCAbdfcepqq6tVV\ndWFVfW6q7YZVdVJVfWn4vvPQXlX1kqo6u6o+W1W3nzrm4GH/L1XVwfOqFwAAYDnzHJ73D0nus0nb\nkUlO7u59kpw8rCfJfZPsM3wdluTlySRkJTkqyR2T7J/kqKWgBQAAsBbmFpq6+yNJLt6k+aAkxw7L\nxyZ58FT7a3vi40l2qqqbJrl3kpO6++LuviTJSfnpIAYAADA3az0RxE26+4Jh+etJbjIs75bk3Kn9\nzhvaVmoHAABYE+s2e153d5LeUuerqsOq6rSqOu2iiy7aUqcFAAC2cWsdmr4xDLvL8P3Cof38JHtM\n7bf70LZS+0/p7qO7e7/u3m/XXXfd4oUDAADbprUOTSckWZoB7+Akx0+1P3qYRe+AJJcOw/jel+Re\nVbXzMAHEvYY2AACANTG3z2mqqjcluVuSXarqvExmwXtukuOq6tAkX03y8GH3E5PcL8nZSb6f5JAk\n6e6Lq+pZSU4d9ntmd286uQQAAMDczC00dfcjV9h0j2X27SSHr3CeVyd59RYsDQAAYGbrNhEEAADA\nRiA0AQAAjBCaAAAARghNAAAAI4QmAACAEUITAADACKEJAABghNAEAAAwQmgCAAAYITQBAACMEJoA\nAABGCE0AAAAjhCYAAIARQhMAAMAIoQkAAGCE0AQAADBCaAIAABghNAEAAIwQmgAAAEYITQAAACOE\nJgAAgBFCEwAAwAihCQAAYITQBAAAMEJoAgAAGCE0AQAAjBCaAAAARghNAAAAI4QmAACAEUITAADA\nCKEJAABghNAEAAAwQmgCAAAYsf16F8DVs9eR7xndfs5z779GlQAAwNZNTxMAAMAIoQkAAGCE0AQA\nADBCaAIAABghNAEAAIwQmgAAAEYITQAAACOEJgAAgBFCEwAAwIjt17sA5mOvI9+z4rZznnv/NawE\nAAA2Nj1NAAAAI4QmAACAEUITAADACKEJAABghNAEAAAwQmgCAAAYITQBAACMEJoAAABGCE0AAAAj\nhCYAAIARQhMAAMAIoQkAAGCE0AQAADBCaAIAABghNAEAAIwQmgAAAEZsv94FsPb2OvI9K24757n3\nX8NKAABg8elpAgAAGCE0AQAAjBCaAAAARnimif/G804AAPDf6WkCAAAYITQBAACMMDyPmRm6BwDA\ntkhPEwAAwAihCQAAYITQBAAAMEJoAgAAGGEiCLaIsUkiVmMSCQAAFpmeJgAAgBFCEwAAwAjD89iw\nfG4UAABrQWhi3Qk/AAAssg0TmqrqPklenGS7JK/q7ueuc0msgZ9lggkAANgSNkRoqqrtkrwsyT2T\nnJfk1Ko6obs/v76VsRHNK4jpFQMA2DptiNCUZP8kZ3f3V5Kkqt6c5KAkQhPLWo8eKsMMAQC2Thsl\nNO2W5Nyp9fOS3HGdaoHNtrUMM/xZwt9ah8qNFGI3Uq0AsC3aKKFpVVV1WJLDhtXvVtUX17OeKbsk\n+eZ6F8HV5v5NqedtrPNmhfs3x+ttcRup1jnw39/G5v5tbO7fxub+ze7ms+y0UULT+Un2mFrffWj7\nse4+OsnRa1nULKrqtO7eb73r4Opx/zY2929jc/82NvdvY3P/Njb3b8vbKB9ue2qSfapq76q6VpJH\nJDlhnWsCAAC2ARuip6m7r6iq/z/J+zKZcvzV3X3WOpcFAABsAzZEaEqS7j4xyYnrXcfVsHBDBtks\n7t/G5v5tbO7fxub+bWzu38bm/m1h1d3rXQMAAMDC2ijPNAEAAKwLoWlOquo+VfXFqjq7qo5c73r4\naVX16qq6sKo+N9V2w6o6qaq+NHzfeWivqnrJcD8/W1W3X7/KSZKq2qOqPlRVn6+qs6rqiKHdPdwA\nquo6VfXJqvrMcP+eMbTvXVWfGO7TW4bJf1JV1x7Wzx6277We9TNRVdtV1RlV9e5h3f3bIKrqnKo6\ns6o+XVWnDW1+f24QVbVTVb21qv6tqr5QVXdy/+ZLaJqDqtouycuS3DfJvkkeWVX7rm9VLOMfktxn\nk7Yjk5zc3fskOXlYTyb3cp/h67AkL1+jGlnZFUme0N37JjkgyeHDf2fu4cZweZIDu/s2SW6b5D5V\ndUCS5yV5YXffMsklSQ4d9j80ySVD+wuH/Vh/RyT5wtS6+7ex3L27bzs1NbXfnxvHi5P8U3ffKslt\nMvnv0P2bI6FpPvZPcnZ3f6W7f5jkzUkOWuea2ER3fyTJxZs0H5Tk2GH52CQPnmp/bU98PMlOVXXT\ntamU5XT3Bd39qWH5skz+h7Fb3MMNYbgP3x1Wrzl8dZIDk7x1aN/0/i3d17cmuUdV1RqVyzKqavck\n90/yqmG94v5tdH5/bgBVdYMkd01yTJJ09w+7+9tx/+ZKaJqP3ZKcO7V+3tDG4rtJd18wLH89yU2G\nZfd0gQ1DfW6X5BNxDzeMYWjXp5NcmOSkJF9O8u3uvmLYZfoe/fj+DdsvTXKjta2YTbwoyZOSXDWs\n3yju30bSSd5fVadX1WFDm9+fG8PeSS5K8ppheOyrqmqHuH9zJTTBCnoytaTpJRdcVV0/yduSPL67\nvzO9zT1cbN19ZXffNsnumfTQ32qdS2JGVfWAJBd29+nrXQtX2126+/aZDN06vKruOr3R78+Ftn2S\n2yd5eXffLsn38pOheEncv3kQmubj/CR7TK3vPrSx+L6x1GU9fL9waHdPF1BVXTOTwPSG7n770Owe\nbjDDsJIPJblTJsNGlj5DcPoe/fj+DdtvkORba1wqP3HnJA+qqnMyGYJ+YCbPWLh/G0R3nz98vzDJ\nOzL5hwu/PzeG85Kc192fGNbfmkmIcv/mSGiaj1OT7DPMInStJI9IcsI618RsTkhy8LB8cJLjp9of\nPcxAc0CSS6e6wFkHw/MQxyT5Qne/YGqTe7gBVNWuVbXTsHzdJPfM5Lm0DyV56LDbpvdv6b4+NMkH\n2wcNrpvufkp3797de2Xy/7gPdvej4v5tCFW1Q1XtuLSc5F5JPhe/PzeE7v56knOr6peGpnsk+Xzc\nv7ny4bZzUlX3y2S893ZJXt3dz1nnkthEVb0pyd2S7JLkG0mOSvLOJMcl2TPJV5M8vLsvHv5Af2km\ns+19P8kh3X3aetTNRFXdJclHk5yZnzxT8dRMnmtyDxdcVd06kweVt8vkH/CO6+5nVtUtMum5uGGS\nM5L8XndfXlXXSfK6TJ5duzjJI7r7K+tTPdOq6m5JntjdD3D/NobhPr1jWN0+yRu7+zlVdaP4/bkh\nVNVtM5mE5VpJvpLkkAy/S+P+zYXQBAAAMMLwPAAAgBFCEwAAwAihCQAAYITQBAAAMEJoAgAAGCE0\nAbAmqqqr6vlT60+sqqdv5jm+u5n7n1NVZw5fn6+qZw/TXwPAzIQmANbK5UkeUlW7rPF1797dv5Zk\n/yS3SPKKLXHSqtp+S5wHgMUnNAGwVq5IcnSSP910Q1XtVVUfrKrPVtXJVbXn0L53Vf3r0FP07E2O\n+d9VdepwzDNWu3h3fzfJHyd5cFXdcOwcVfUXVfXFqvpYVb2pqp44tJ9SVS+qqtOSHFFVu1bV24Zz\nnFpVdx7226GqXl1Vn6yqM6rqoKv9UwNg3QlNAKyllyV5VFXdYJP2v0tybHffOskbkrxkaH9xkpcP\nPUUXLO1cVfdKsk8mvUe3TfI/ququq128u7+T5D+S7LPSOarqDkl+O8ltktw3yX6bnOZa3b1fdz9/\nqO+F3b10zKuGff48yQe7e/8kd0/yt1W1w2r1AbCYDC0AYM1093eq6rVJ/iTJf01tulOShwzLr0vy\nN8PynTMJI0vtzxuW7zV8nTGsXz+TAPSRGcqoVc6xY5Lju/sHSX5QVe/a5Pi3TC3/VpJ9q5ZOmZ+r\nqusP533QUg9Vkusk2TPJF2aoD4AFIzQBsNZelORTSV4z4/69TFsl+evu3qznk6pqxyR7Jfn3lc5R\nVY9f5TTfm1q+RpIDhoA1fY5K8tvd/cXNqQ+AxWR4HgBrqrsvTnJckkOnmv8lySOG5Ucl+eiw/M+b\ntC95X5I/HHp1UlW7VdWNx6477Pv3Sd7Z3ZeMnOOfkzywqq4zbHvAyGnfn+RxU9e47VR9jxvCU6rq\ndmO1AbDYhCYA1sPzk0zPove4JIdU1WeT/H6SI4b2I5IcXlVnJtltaefufn+SNyb512HbWzMZVrec\nD1XV55J8Msl/Jnns2Dm6+9QkJyT5bJL3JjkzyaUrnPtPkuw3TCTx+UwmmkiSZyW5ZpLPVtVZwzoA\nG1R1LzfqAQC2XVV1/e7+blVdL5PnpA7r7k+td10ArA/PNAHATzu6qvbNZAKHYwUmgG2bniYAAIAR\nnmkCAAAYITQBAACMEJoAAABGCE0AAAAjhCYAAIARQhMAAMCI/wf6RBPKPECezgAAAABJRU5ErkJg\ngg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fe395e1fb00>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plot node degree to check if distribution follows power law\n",
    "node_degrees = ppi_network.sum()\n",
    "fig = plt.figure(figsize=(14, 8))\n",
    "bins = np.linspace(1,node_degrees.max(), 100)\n",
    "ax = plt.hist(node_degrees, bins)\n",
    "plt.xlabel('Node Degree')\n",
    "plt.ylabel('Number of genes with node degree')\n",
    "plt.title('Node Degree Distribution of Genes in PPI')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## STRING DB: Get PPI network\n",
    "I also want to construct the adjacency matrix for the STRING DB PPI network because it seems as if ConsensusPathDB is lacking some interactions I am interested in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-63-552bd6f7b850>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0minteractions_nona\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minteractions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0minteractions_nona\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotein1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minteractions_nona\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotein1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'9606.'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m \u001b[0minteractions_nona\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotein2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minteractions_nona\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotein2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'9606.'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;31m# plot distribution of scores\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/pkg/python-3.6.0-0/lib/python3.6/site-packages/pandas/core/strings.py\u001b[0m in \u001b[0;36mreplace\u001b[0;34m(self, pat, repl, n, case, flags)\u001b[0m\n\u001b[1;32m   1504\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrepl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcase\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1505\u001b[0m         result = str_replace(self._data, pat, repl, n=n, case=case,\n\u001b[0;32m-> 1506\u001b[0;31m                              flags=flags)\n\u001b[0m\u001b[1;32m   1507\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_wrap_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1508\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/pkg/python-3.6.0-0/lib/python3.6/site-packages/pandas/core/strings.py\u001b[0m in \u001b[0;36mstr_replace\u001b[0;34m(arr, pat, repl, n, case, flags)\u001b[0m\n\u001b[1;32m    334\u001b[0m         \u001b[0mf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrepl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    335\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 336\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_na_map\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    337\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    338\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/pkg/python-3.6.0-0/lib/python3.6/site-packages/pandas/core/strings.py\u001b[0m in \u001b[0;36m_na_map\u001b[0;34m(f, arr, na_result, dtype)\u001b[0m\n\u001b[1;32m    152\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0m_na_map\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_result\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnan\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    153\u001b[0m     \u001b[0;31m# should really _check_ for NA\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 154\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_map\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_mask\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_value\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mna_result\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    155\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    156\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/pkg/python-3.6.0-0/lib/python3.6/site-packages/pandas/core/strings.py\u001b[0m in \u001b[0;36m_map\u001b[0;34m(f, arr, na_mask, na_value, dtype)\u001b[0m\n\u001b[1;32m    167\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    168\u001b[0m             \u001b[0mconvert\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmask\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 169\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap_infer_mask\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mview\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0muint8\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    170\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mTypeError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mAttributeError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    171\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/src/inference.pyx\u001b[0m in \u001b[0;36mpandas.lib.map_infer_mask (pandas/lib.c:65674)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m/pkg/python-3.6.0-0/lib/python3.6/site-packages/pandas/core/strings.py\u001b[0m in \u001b[0;36mf\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m    330\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    331\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 332\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mregex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msub\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrepl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcount\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    333\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    334\u001b[0m         \u001b[0mf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrepl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# load data\n",
    "interactions = pandas.read_csv('../data/networks/9606_stringdb_v10_5.txt.gz',\n",
    "                                         header=0,\n",
    "                                         sep=' ',\n",
    "                                         encoding='utf8'\n",
    "                                        )\n",
    "# remove NAs and the trailing species name\n",
    "interactions_nona = interactions.dropna()\n",
    "interactions_nona.protein1 = interactions_nona.protein1.str.replace('9606.', '')\n",
    "interactions_nona.protein2 = interactions_nona.protein2.str.replace('9606.', '')\n",
    "\n",
    "# plot distribution of scores\n",
    "fig = plt.figure(figsize=(14, 8))\n",
    "ax = interactions_nona.combined_score.plot.hist(bins=50, title='Distribution of Scores')\n",
    "plt.xlabel('Combined Scores (STRING DB)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-20...done.\n",
      "querying 1-20...done.\n"
     ]
    }
   ],
   "source": [
    "# use bioMart to convert protein and gene IDs\n",
    "mg = mygene.MyGeneInfo()\n",
    "f = lambda x: x['ensembl'][0]['gene'] if type(x['ensembl']) is list else x['ensembl']['gene']\n",
    "res = mg.getgenes(interactions_nona.protein1, 'symbol,ensembl.gene', scopes='ensembl.protein')\n",
    "proteins1 = [f(x) for x in res]\n",
    "res = mg.getgenes(interactions_nona.protein2, 'symbol,ensembl.gene', scopes='ensembl.protein')\n",
    "proteins2 = [f(x) for x in res]\n",
    "\n",
    "# put into dataframe\n",
    "interactions_nona['gene1'] = proteins1\n",
    "interactions_nona['gene2'] = proteins2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000263431</td>\n",
       "      <td>260</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000126583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000353863</td>\n",
       "      <td>164</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000163848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000342026</td>\n",
       "      <td>159</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000117592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000240874</td>\n",
       "      <td>194</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000160145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000379847</td>\n",
       "      <td>164</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000166478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000400088</td>\n",
       "      <td>189</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000250506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000362465</td>\n",
       "      <td>240</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000102383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000415070</td>\n",
       "      <td>164</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000270011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000379387</td>\n",
       "      <td>164</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000283034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000335632</td>\n",
       "      <td>224</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000187446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000356143</td>\n",
       "      <td>255</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000170382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000284957</td>\n",
       "      <td>320</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000154710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000332668</td>\n",
       "      <td>255</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000182704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000371734</td>\n",
       "      <td>190</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000107104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000350331</td>\n",
       "      <td>190</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000196372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000251412</td>\n",
       "      <td>266</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000037042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000270451</td>\n",
       "      <td>164</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000171425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000301744</td>\n",
       "      <td>164</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000167981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000269571</td>\n",
       "      <td>164</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000141736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000471417</td>\n",
       "      <td>190</td>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENSG00000189045</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           protein1         protein2  combined_score            gene1  \\\n",
       "0   ENSP00000000233  ENSP00000263431             260  ENSG00000004059   \n",
       "1   ENSP00000000233  ENSP00000353863             164  ENSG00000004059   \n",
       "2   ENSP00000000233  ENSP00000342026             159  ENSG00000004059   \n",
       "3   ENSP00000000233  ENSP00000240874             194  ENSG00000004059   \n",
       "4   ENSP00000000233  ENSP00000379847             164  ENSG00000004059   \n",
       "5   ENSP00000000233  ENSP00000400088             189  ENSG00000004059   \n",
       "6   ENSP00000000233  ENSP00000362465             240  ENSG00000004059   \n",
       "7   ENSP00000000233  ENSP00000415070             164  ENSG00000004059   \n",
       "8   ENSP00000000233  ENSP00000379387             164  ENSG00000004059   \n",
       "9   ENSP00000000233  ENSP00000335632             224  ENSG00000004059   \n",
       "10  ENSP00000000233  ENSP00000356143             255  ENSG00000004059   \n",
       "11  ENSP00000000233  ENSP00000284957             320  ENSG00000004059   \n",
       "12  ENSP00000000233  ENSP00000332668             255  ENSG00000004059   \n",
       "13  ENSP00000000233  ENSP00000371734             190  ENSG00000004059   \n",
       "14  ENSP00000000233  ENSP00000350331             190  ENSG00000004059   \n",
       "15  ENSP00000000233  ENSP00000251412             266  ENSG00000004059   \n",
       "16  ENSP00000000233  ENSP00000270451             164  ENSG00000004059   \n",
       "17  ENSP00000000233  ENSP00000301744             164  ENSG00000004059   \n",
       "18  ENSP00000000233  ENSP00000269571             164  ENSG00000004059   \n",
       "19  ENSP00000000233  ENSP00000471417             190  ENSG00000004059   \n",
       "\n",
       "              gene2  \n",
       "0   ENSG00000126583  \n",
       "1   ENSG00000163848  \n",
       "2   ENSG00000117592  \n",
       "3   ENSG00000160145  \n",
       "4   ENSG00000166478  \n",
       "5   ENSG00000250506  \n",
       "6   ENSG00000102383  \n",
       "7   ENSG00000270011  \n",
       "8   ENSG00000283034  \n",
       "9   ENSG00000187446  \n",
       "10  ENSG00000170382  \n",
       "11  ENSG00000154710  \n",
       "12  ENSG00000182704  \n",
       "13  ENSG00000107104  \n",
       "14  ENSG00000196372  \n",
       "15  ENSG00000037042  \n",
       "16  ENSG00000171425  \n",
       "17  ENSG00000167981  \n",
       "18  ENSG00000141736  \n",
       "19  ENSG00000189045  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interactions_nona"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "not found ENSG00000004059 -> ENSG00000126583\n",
      "not found ENSG00000004059 -> ENSG00000283034\n"
     ]
    }
   ],
   "source": [
    "INTERACTION_SCORE_THRESHOLD = 1\n",
    "\n",
    "# extract adjacency matrix from interactions\n",
    "N = len(gene_names)\n",
    "adjacency_matrix = np.zeros((N, N), np.uint8)\n",
    "adj_df = pandas.DataFrame(adjacency_matrix, index=gene_ids, columns=gene_ids)\n",
    "\n",
    "count = 0\n",
    "for index, row in interactions_nona.iterrows():\n",
    "    if row.combined_score > INTERACTION_SCORE_THRESHOLD:\n",
    "        i1 = row.gene1\n",
    "        i2 = row.gene2\n",
    "        if i1 in gene_ids and i2 in gene_ids:\n",
    "            adj_df.ix[i1, i2] = 1\n",
    "            adj_df.ix[i2, i1] = 1\n",
    "        else:\n",
    "            print (\"not found {} -> {}\".format(i1, i2))\n",
    "            \n",
    "    count += 1\n",
    "    if count % 10000 == 0:\n",
    "        print (\"Processed {} out of {} rows\".format(count, interactions_nona.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "634.0"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ppi_network.sum().max()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Remove isolated genes from network & gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9624, 9624)\n"
     ]
    }
   ],
   "source": [
    "# kick out all genes without interactions (isolated nodes)\n",
    "genes_with_interactions = ppi_network.sum(axis=1) > 0\n",
    "ppi_network = ppi_network.ix[genes_with_interactions, genes_with_interactions]\n",
    "print (ppi_network.shape)\n",
    "\n",
    "# build list of gene names and ids\n",
    "gene_names_ids = ppi_network.join(gene_expression['Name']).Name\n",
    "gene_names_ids = pandas.DataFrame(gene_names_ids)\n",
    "gene_names_ids = np.vstack([gene_names_ids.index, gene_names_ids.Name])\n",
    "gene_names_ids = gene_names_ids.transpose(1, 0)\n",
    "\n",
    "# remove gene expression for isolated genes\n",
    "gene_expression_short = ge_final[ge_final.index.isin(ppi_network.index)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Get raw counts only for network genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ppi_network' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-29-6b93b4c3cc44>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;31m# only use genes for which we have expression\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m \u001b[0mcounts_with_ge\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcounts\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcounts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mppi_network\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m \u001b[0mcounts_with_ge\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'ppi_network' is not defined"
     ]
    }
   ],
   "source": [
    "counts = pandas.DataFrame.from_csv('../data/differential_expression/readcounts_raw.tsv', sep='\\t', index_col=2)\n",
    "\n",
    "# call the column with gene name in it 'Name'\n",
    "new_cols = counts.columns.values\n",
    "new_cols[1] = 'Name'\n",
    "counts.columns = new_cols\n",
    "\n",
    "# rearrange columns to have name first, then the values\n",
    "cols = counts.columns.tolist()\n",
    "cols = [cols[1]] + [cols[0]] + cols[2:]\n",
    "counts = counts[cols]\n",
    "\n",
    "# only use genes for which we have expression\n",
    "counts_with_ge = counts[counts.index.isin(ppi_network.index)]\n",
    "counts_with_ge.head()\n",
    "\n",
    "# rename columns to have correct names (not some .bam file)\n",
    "counts_with_ge.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Get Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9624, 24)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load the labels\n",
    "infection_genes = pd.DataFrame.from_csv('../data/labels/labels_wigo_codingonly.csv', sep='\\t', index_col=3)\n",
    "my_infection_genes = pd.DataFrame.from_csv('../data/labels/my_labels_literaturesearch.tsv',\n",
    "                                           sep='\\t',\n",
    "                                           index_col=1\n",
    "                                          )\n",
    "infection_genes = infection_genes.append(my_infection_genes)\n",
    "\n",
    "# do train test splits\n",
    "# Construct DF with label column and gene expression\n",
    "gene_names_df = pd.DataFrame(gene_names,\n",
    "                             index=gene_names[:, 0],\n",
    "                             columns=['ID', 'name']\n",
    "                            ).drop('ID', axis=1)\n",
    "gene_names_df['node_number'] = np.arange(0, gene_names_df.shape[0])\n",
    "ge = gene_expression_data.set_index(gene_names_df.index)\n",
    "features_labeled = ge.join(gene_names_df)\n",
    "assert (features_labeled.isnull().sum().sum() == 0)\n",
    "features_labeled['label'] = gene_names_df.index.isin(ens_ids)\n",
    "\n",
    "# split for training, testing & validation\n",
    "X_train, X_test = train_test_split(features_labeled,\n",
    "                                   stratify=features_labeled.label,\n",
    "                                   test_size=TEST_RATIO\n",
    "                                  )\n",
    "X_val = X_train[-VAL_SIZE:]\n",
    "X_train = X_train[:-VAL_SIZE]\n",
    "\n",
    "print (\"Split Training and Testing with {}% test nodes\".format(TEST_RATIO*100.))\n",
    "print (\"Training Nodes: {}\\t# of Labels in Train Set: {}\".format(X_train.shape[0], X_train.label.sum()))\n",
    "print (\"Testing Nodes: {}\\t# of Labels in Test Set: {}\".format(X_test.shape[0], X_test.label.sum()))\n",
    "\n",
    "# construct training, testing and validation masks\n",
    "def build_mask(features_labeled, X):\n",
    "    mask = features_labeled[features_labeled.isin(X)].label\n",
    "    mask[mask.isnull()] = 0\n",
    "    # sanity check\n",
    "    assert (np.all(mask.index == gene_names[:, 0]))\n",
    "    return mask.values\n",
    "\n",
    "train_mask = build_mask(features_labeled, X_train)\n",
    "test_mask = build_mask(features_labeled, X_test)\n",
    "val_mask = build_mask(features_labeled, X_val)\n",
    "\n",
    "# construct labels\n",
    "y_train = pd.get_dummies(features_labeled[features_labeled.isin(X_train)].label).values\n",
    "y_test = pd.get_dummies(features_labeled[features_labeled.isin(X_test)].label).values\n",
    "y_val = pd.get_dummies(features_labeled[features_labeled.isin(X_val)].label).values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Generate Training, Testing & Validation Splits"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Write to hdf5 File on Disk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# create gene names\n",
    "string_dt = h5py.special_dtype(vlen=str)\n",
    "\n",
    "f = h5py.File('../data/preprocessing/ppi_networks.h5', 'w')\n",
    "f.create_dataset('consensusPathDB_ppi', data=ppi_network, shape=ppi_network.shape)\n",
    "f.create_dataset('gene_expression', data=gene_expression_short, shape=gene_expression_short.shape)\n",
    "f.create_dataset('gene_names', data=gene_names_ids, dtype=string_dt)\n",
    "\n",
    "# we don't have labels yet :-(\n",
    "#f.create_dataset('y_train', data=y_train, shape=y_train.shape)\n",
    "#f.create_dataset('y_test', data=y_test, shape=y_test.shape)\n",
    "#f.create_dataset('y_val', data=y_val, shape=y_val.shape)\n",
    "\n",
    "#f.create_dataset('mask_train', data=mask_train, shape=mask_train.shape)\n",
    "#f.create_dataset('mask_test', data=mask_test, shape=mask_test.shape)\n",
    "#f.create_dataset('mask_val', data=mask_val, shape=mask_val.shape)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
