{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Preprocess Legionella Data & PPI to Form GCN Input\n",
    "In this notebook, I want to join the PPI network from ConsensusPathDB, the gene expression derived from the dual RNA-seq experiments and the labels to form one dataset. The goal is a hdf5 container with all the information to train a GCN on the data.\n",
    "\n",
    "This involves the following steps:\n",
    "* Preprocess the gene expression data\n",
    "* Calculate an adjacency matrix from the protein-protein-interactions\n",
    "* Only use genes in network for which we have gene expression\n",
    "* Check if the node degree follows a power law distribution\n",
    "* Load the labels and split into training and testing\n",
    "* Create hdf5 container"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sasse/.local/lib/python3.6/site-packages/h5py/__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.\n",
      "  from ._conv import register_converters as _register_converters\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "#from sklearn.decomposition import PCA\n",
    "from sklearn import preprocessing\n",
    "import matplotlib.pyplot as plt\n",
    "np.set_printoptions(suppress=True)\n",
    "import h5py\n",
    "import mygene\n",
    "from sklearn.model_selection import train_test_split\n",
    "import networkx as nx\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Parameters for the Notebook\n",
    "* **TIMES_STD_THRESHOLD:** How many times over the standard deviation does a gene have to show expression to be considered as an outlier.\n",
    "* **MAX_ZEROS_ALLOWED:** The ratio of zeros of gene expressions for a gene that is permitted before the gene gets kicked out.\n",
    "* **SCALE:** Do we normalize the gene expression using a standard min/max scaler?\n",
    "* **VAL_SIZE:** The size of our validation data set (0 means no validation data set).\n",
    "* **TEST_RATIO:** The ratio of the data that is used for testing the model.\n",
    "* **BALANCE:** Do we want the training set to contain as many positives as negatives or not?\n",
    "* **MIN_VAR_THRESHOLD:** When we derive a set of negative genes (those definitely not involved in Legionella infection), we select them using the genes which show low variance over the course of infection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# params\n",
    "TIMES_STD_THRESHOLD = 10.\n",
    "MAX_ZEROS_ALLOWED = .7\n",
    "SCALE = False\n",
    "VAL_SIZE = 0\n",
    "TEST_RATIO = 0.4\n",
    "BALANCE = False\n",
    "MIN_VAR_THRESHOLD = 0.5\n",
    "NEGATIVE_SET_LOWVAR = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Preprocess Gene Expression\n",
    "Let's start by preprocessing the gene expression values. We have the counts already normalized, which is good. But we might want to kick out those genes that show a unrealistically high gene expression or that do contain lots of zeros."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Had gene expression for 41424 genes in the beginning.\n",
      "kicked out 9 super highly expressed genes\n",
      "Kicked out 11344 genes with more than 70.0% zeros\n",
      "==> Left with gene expression for 30071 genes\n"
     ]
    }
   ],
   "source": [
    "# load data\n",
    "gene_expression = pd.DataFrame.from_csv('../data/preprocessing/normalized-counts-labels.csv',\n",
    "                                        encoding='utf-8', sep=','\n",
    "                                       )\n",
    "ge_nonames = gene_expression.drop('Name', axis=1)\n",
    "\n",
    "# kick out super highly expressed genes\n",
    "threshold = ge_nonames.mean(axis=1).std()*TIMES_STD_THRESHOLD\n",
    "#print (\"Threshold Gene Expression: {}\".format(threshold))\n",
    "anomalies = gene_expression[ge_nonames.mean(axis=1) > threshold]\n",
    "ge_anomalies_removed = ge_nonames.drop(anomalies.index)\n",
    "\n",
    "# kick out genes with too many zeros\n",
    "ge_zeros_removed = ge_anomalies_removed[ge_anomalies_removed.astype('bool').mean(axis=1)>=(1-MAX_ZEROS_ALLOWED)]\n",
    "\n",
    "# write gene expression to file to perform differential expression analysis on it\n",
    "ge_zeros_removed.to_csv('../data/preprocessing/gene_expression_counts.tsv', sep='\\t')\n",
    "\n",
    "# scaling\n",
    "if SCALE:\n",
    "    scaler = preprocessing.StandardScaler()\n",
    "    scaled_features = scaler.fit_transform(ge_zeros_removed)\n",
    "    ge_final = pd.DataFrame(scaled_features,\n",
    "                            index=ge_zeros_removed.index,\n",
    "                            columns=ge_zeros_removed.columns\n",
    "                           )\n",
    "else:\n",
    "    ge_final = ge_zeros_removed\n",
    "\n",
    "# get the name column back\n",
    "ge_final = gene_expression[gene_expression.index.isin(ge_final.index)]\n",
    "\n",
    "# print some information\n",
    "print (\"Had gene expression for {} genes in the beginning.\".format(ge_nonames.shape[0]))\n",
    "print (\"kicked out {} super highly expressed genes\".format(ge_nonames.shape[0] - ge_anomalies_removed.shape[0]))\n",
    "print (\"Kicked out {} genes with more than {}% zeros\".format(ge_anomalies_removed.shape[0]-ge_zeros_removed.shape[0],\n",
    "                                                             MAX_ZEROS_ALLOWED*100.\n",
    "                                                            ))\n",
    "print (\"==> Left with gene expression for {} genes\".format(ge_final.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## ConsensusPathDB: Get PPI Network\n",
    "I want to construct an adjacency matrix from the interactions in ConsensusPathDB.\n",
    "However, I only want it for the genes for which I have gene expression data. If not both genes are involved in an interaction, I don't want to save the interaction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We lose 53664 interactions by removing those where not both genes have expression\n",
      "Still keeping 164708 interactions\n"
     ]
    }
   ],
   "source": [
    "# read interaction data (implicit threshold of 0.5 for interaction confidence)\n",
    "interactions = pd.read_csv('../data/networks/CPDB_ensg_edgelist.tsv',\n",
    "                           header=0,\n",
    "                           sep='\\t',\n",
    "                           names=['number', 'gene1', 'gene2', 'score']\n",
    "                          )\n",
    "\n",
    "# kick out interactions for which we don't have gene expression\n",
    "lost = interactions[~interactions.gene1.isin(ge_final.index) | ~interactions.gene2.isin(ge_final.index)].shape[0]\n",
    "print (\"We lose {} interactions by removing those where not both genes have expression\".format(lost))\n",
    "interactions_ge = interactions[interactions.gene1.isin(ge_final.index) & interactions.gene2.isin(ge_final.index)]\n",
    "print (\"Still keeping {} interactions\".format(interactions_ge.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed Adjacency Matrix with average node degree of: 30.425020789060333\n",
      "Adjacency matrix has 329290.0 edges in total\n",
      "Network has 10823 nodes in total\n"
     ]
    }
   ],
   "source": [
    "# construct adjacency matrix (memory consuming)\n",
    "G = nx.from_pandas_dataframe(interactions_ge, source='gene1', target='gene2', edge_attr=None)\n",
    "ppi_network = nx.to_pandas_dataframe(G)\n",
    "\n",
    "# sanity check (matrix symmetric?)\n",
    "assert (np.allclose(ppi_network, ppi_network.T))\n",
    "\n",
    "# output some statistics\n",
    "avg_node_degree = ppi_network.sum().mean()\n",
    "total_num_of_connections = ppi_network.sum().sum()\n",
    "\n",
    "print (\"Constructed Adjacency Matrix with average node degree of: {}\".format(avg_node_degree))\n",
    "print (\"Adjacency matrix has {} edges in total\".format(total_num_of_connections))\n",
    "print (\"Network has {} nodes in total\".format(ppi_network.index.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x7f272f170400>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA00AAAHwCAYAAAB6yISuAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xu8bWVZN/zfJXhEEhA05SCUdNAeTw8KhpXns+JjZpoV\nIYW9L69hWYlHPKeZmqd8RFHRTCWPW8UUD2hWKiAmHjIQUUAUFATURJHr/WOOpdPtWmPNvdlzrbn3\n/n4/n/VZY9xjjHtec405Yf72PcY9q7sDAADA8q6x3gUAAAAsMqEJAABghNAEAAAwQmgCAAAYITQB\nAACMEJoAAABGCE0AW1BVPbWq/nG961hUVfXeqjp0C/X1G1X1xan1c6rq7lui76G/z1XVnbdUfzM+\nZlXVa6rqkqr65Fo+9pZUVU+oqletdx0AW4rQBGzXhg/aF1bVTlNtf1xVJ69TLf9TVZdX1ber6t+r\n6k+raqv4b3VVdVV9t6q+U1XfqqoPVtXvTu/T3ffp7uNn7OvmY/t097929y9f3bqHx3ttVT1zo/5v\n2d0nb4n+N8GdktwjyV7dfYfldqiqm1TVK6vqa8Pf+uyh/l9Z21JX1t3P7u4/3pxjh+fyg+G5XVxV\nJy09t+EfJX44bFt6j9xx2PZHVfWxLfk8AJZsFf8jBpizHZIctd5FDB7Q3TsnuVmS5yR5XJLj5vFA\nVbXDHLq9dXdfP8kvJ3ltkpdW1TFb+kGqasct3eeCuFmSc7r7u8ttrKobJvn3JNdL8htJdk5yuyQf\nySRsbSv+dngd7ZXkwkxeS0vePGzbI8nHkrytqmrtSwS2J0ITQPK8JH9ZVbsst7Gqfr2qTqmqS4ff\nvz61bb+q+sgwOnRSkt03Ovag4V/Dv11V/znr5V7dfWl3b0jyu0kOrapfG/q7dlX9XVV9taq+UVX/\nt6quO/V4f11VFwyjEH88PWIz/Av+y6vqxKr6bpK7zNDf/avq01P/qn+rGev/Zne/Psn/k+Txw4f9\nVNXJVfXHw/LNh7/dpVX1zap689D+0aGb/xxGFH63qu5cVedV1eOq6utJXrPUttFD376qPj9c3vaa\nqrrO0OfPjEIs/W2q6ogkj0jy18PjvWvY/uPL/Ya/098Pf9evDcvXHrYt1fbYmoxaXlBVh630t6mq\nm1bVhmEU5ayq+pOh/fAkr0pyx6GOpy1z+J8nuSzJH3T3l3ri2939mu5+ydRjrPi6G87BM6rq34bX\n7furavcZj/2jmoxsXV5VX66qR6zwHH98mWpV7Tv8rQ8dXmffrKonrvT3mdbd30vyT0l+bZltP0xy\nfJKfT3LDWfoD2FxCE0ByapKTk/zlxhuqarck70ny4kw+mL0gyXuWQkAmH+hOyyQsPSPJoVPH7jkc\n+8wkuw39v7Wq9pi1sO7+ZJLzMhlVSCajT7+U5DZJbp5kzyRPGR7v3kn+Isndh213XqbL30vyrExG\nKD62Sn+3TfLqJI8anvsrkmxYCgszemeSHZMsd6nZM5K8P8mumYwovGR4zr85bL91d1+/u988rP98\nJn/HmyU5YoXHe0SSeyX5xeF5PWm1Arv72CRvyDC60d0PWGa3JyY5KJO/062H5zPd988nuUEmf7/D\nk7ysqnZd4SHflMk5vWmShyR5dlXdtbuPS/KnSf5jqGO5Ebq7J3l7d1+10vOZ8XX3e0kOS3KjJNca\n9hk9tiaXsL44yX2G0dBfT/LplepYxp0yGYG8W5KnVNWvrnZAVV0/k3N6+jLbrp3kj5Kc293f3IQ6\nADaZ0AQw8ZQkj14m0NwvyZnd/fruvrK735jkv5I8oKr2SXL7JE/u7iu6+6NJ3jV17O8nObG7T+zu\nq7r7pEwC2n03sbavJdmtqiqTsPDn3X1xd1+e5NlJHjbs99Akr+nuzw3/Qv/UZfp6Z3f/2/Ch+4pV\n+jsiySu6+xPd/aPhXqQrMgkPMxlGA76ZyQfwjf0wkwB00+7+fnevdj/KVUmOGf7W/7PCPi/t7nO7\n++JMwuHDZ611FY9I8vTuvrC7L0rytCR/MLX9h8P2H3b3iUm+k0lA+ClVtXeSg5M8bnjOn85kdOkP\nZ6xj9yRfn+rvgcOI0OVV9f6heZbX3Wu6+7+Hv+MJmYTBWY69KsmvVdV1u/uC7v7cjHUnydO6+3+6\n+z+T/Gcm4XMlf1lV305yVpLrZxKOljx02HZukv+d5P9sQg0Am0VoAkjS3Z9N8u4kR2+06aZJvrJR\n21cyGVG4aZJLNrr/ZHrfmyX5neFD7beHD3p3SnKTTSxvzyQXZ3IPx/WSnDbV378M7Uu1njt13Ln5\nWdNtq/V3sySP3aj+vYfHmUlVXXPo7+JlNv91kkryyZrMVPfIVbq7qLu/v8o+08/vK5tS6yo2fh1s\n3Pe3uvvKqfXvZfJhf7l+lgLqdF97zljHtzL1+unuDd29SyaX7V1raJ7ldff1qeXpWlc8dnid/24m\no2EXVNV7atMmn1jpMZfzd929S3f/fHc/sLu/NLXthGHbjbr7rt192ibUALBZttUbaQE2xzFJPpXk\n+VNtX8vkg+S0fTIJFxck2bWqdpoKTvsk6WH53CSv7+4/2dyCqur2mXyg/lgmIzb/k+SW3X3+Mrtf\nkMllbkv2Xmafnlperb9zkzyru5+1ObUPDklyZZKfmT67u7+eZOl+njsl+UBVfbS7z1qhr16hfdr0\nc94nk/OXJN/NJCBmeLyf38S+l14HSyMr031viqVRw52ngtM+SZb7+y/ng0keVFVPG7lE7+q87kaP\n7e73JXlfTe57e2aSV+Ynl44CbLOMNAEMhg/rb07yZ1PNJyb5par6varasSZTaN8iybu7+yuZXLr0\ntKq61vDBf/p+mH/M5DK+e1XVDlV1nWHSgOlgs6yq+rmqun8m97/8Y3efMXxIfmWSF1bVjYb99qyq\new2HnZDksKr61aq6XpInr/J8V+vvlUn+tKoOrImdqup+VbXzDPXvNkwS8LIkz+3uby2zz+9M/S0u\nySS4LAWBbyT5hdUeZxlHVtVew71oT8zkfCaTy8FuWVW3qcnkEE/d6LjVHu+NSZ403NuzeyaXc27y\n93F197mZzH73N8Pr4VaZ3AM1a18vyOQesNdX1S8O52Xn/OTyuuRqvO7Gjq2qG1fVIcO9TVdkcgni\nivdWAWxLhCaAn/b0JD/+zqbhw/79kzw2k0uj/jrJ/aduPP+9JAdmcvnZMUleN3XsuZmMtDwhyUWZ\n/Cv+X2X8v73vqqrLh32fmMmH5OmZ2B6XyX0eH6+qy5J8IMO9M9393kxu1P/w0j7DMVeMPN5Yf6dm\nMhL00kxCzVn56XtLlvOfVfWdYd8/zuR+qaessO/tk3xi2H9DkqO6++xh21OTHD9cIvbQVR5z2j9l\nMrnE2Um+lMloSLr7vzM5tx9IcmYmI3fTjktyi+Hx3rFMv8/MJCB/JskZmYxIPnOZ/Wbx8CT7ZjLq\n9PZM7tP6wCwHDq+7g5J8P5PncHkmkzHsnMlMhZv7ulvqf+zYa2Qy0cjXMnm9/9bSYwJs66p7lqsd\nANjaDLOTfTbJtTe63wYA2ARGmgC2IVX1f2rynUK7JnlukncJTABw9QhNANuWRyW5MJNL034Ul08B\nwNXm8jwAAIARRpoAAABGCE0AAAAjtskvt91999173333Xe8yAACABXbaaad9s7v3WG2/bTI07bvv\nvjn11FPXuwwAAGCBVdVXZtnP5XkAAAAjhCYAAIARQhMAAMAIoQkAAGCE0AQAADBCaAIAABghNAEA\nAIwQmgAAAEYITQAAACOEJgAAgBFCEwAAwAihCQAAYITQBAAAMEJoAgAAGCE0AQAAjBCaAAAARghN\nAAAAI4QmAACAEUITAADAiB3Xu4Dtwb5Hv2fFbec8535rWAkAALCpjDQBAACMEJoAAABGCE0AAAAj\nhCYAAIARQhMAAMAIoQkAAGCE0AQAADBCaAIAABghNAEAAIwQmgAAAEYITQAAACOEJgAAgBFCEwAA\nwAihCQAAYITQBAAAMEJoAgAAGCE0AQAAjBCaAAAARghNAAAAI4QmAACAEUITAADACKEJAABghNAE\nAAAwQmgCAAAYITQBAACMEJoAAABGCE0AAAAjhCYAAIARQhMAAMAIoQkAAGCE0AQAADBCaAIAABgh\nNAEAAIwQmgAAAEYITQAAACOEJgAAgBFCEwAAwAihCQAAYITQBAAAMEJoAgAAGCE0AQAAjJhraKqq\nc6rqjKr6dFWdOrTtVlUnVdWZw+9dh/aqqhdX1VlV9Zmqut1UP4cO+59ZVYfOs2YAAIBpazHSdJfu\nvk13HzCsH53kg929f5IPDutJcp8k+w8/RyR5eTIJWUmOSXJgkjskOWYpaAEAAMzbelyed0iS44fl\n45M8aKr9dT3x8SS7VNVNktwryUndfXF3X5LkpCT3XuuiAQCA7dO8Q1MneX9VnVZVRwxtN+7uC4bl\nrye58bC8Z5Jzp449b2hbqf2nVNURVXVqVZ160UUXbcnnAAAAbMd2nHP/d+ru86vqRklOqqr/mt7Y\n3V1VvSUeqLuPTXJskhxwwAFbpE8AAIC5jjR19/nD7wuTvD2Te5K+MVx2l+H3hcPu5yfZe+rwvYa2\nldoBAADmbm6hqap2qqqdl5aT3DPJZ5NsSLI0A96hSd45LG9I8ofDLHoHJbl0uIzvfUnuWVW7DhNA\n3HNoAwAAmLt5Xp534yRvr6qlx/mn7v6XqjolyQlVdXiSryR56LD/iUnum+SsJN9LcliSdPfFVfWM\nJKcM+z29uy+eY90AAAA/NrfQ1N1nJ7n1Mu3fSnK3Zdo7yZEr9PXqJK/e0jUCAACsZj2mHAcAANhq\nCE0AAAAjhCYAAIARQhMAAMAIoQkAAGCE0AQAADBCaAIAABghNAEAAIwQmgAAAEYITQAAACOEJgAA\ngBFCEwAAwAihCQAAYITQBAAAMEJoAgAAGCE0AQAAjBCaAAAARghNAAAAI4QmAACAEUITAADACKEJ\nAABghNAEAAAwQmgCAAAYITQBAACMEJoAAABGCE0AAAAjhCYAAIARQhMAAMAIoQkAAGCE0AQAADBC\naAIAABghNAEAAIwQmgAAAEYITQAAACOEJgAAgBFCEwAAwAihCQAAYITQBAAAMEJoAgAAGCE0AQAA\njBCaAAAARghNAAAAI4QmAACAEUITAADACKEJAABghNAEAAAwQmgCAAAYITQBAACMEJoAAABGCE0A\nAAAjhCYAAIARQhMAAMAIoQkAAGCE0AQAADBCaAIAABghNAEAAIwQmgAAAEYITQAAACNmCk1VdbOq\nuvuwfN2q2nm+ZQEAACyGVUNTVf1JkrckecXQtFeSd8yzKAAAgEUxy0jTkUkOTnJZknT3mUluNM+i\nAAAAFsUsoemK7v7B0kpV7Zik51cSAADA4pglNH2kqp6Q5LpVdY8k/5zkXfMtCwAAYDHMEpqOTnJR\nkjOSPCrJiUmeNM+iAAAAFsWqoam7r0ryj0me1N0P6e5XdvfMl+dV1Q5VdXpVvXtY36+qPlFVZ1XV\nm6vqWkP7tYf1s4bt+0718fih/YtVda9NfZIAAACba5bZ8x6Y5NNJ/mVYv01VbdiExzgqyRem1p+b\n5IXdffMklyQ5fGg/PMklQ/sLh/1SVbdI8rAkt0xy7yT/UFU7bMLjAwAAbLZZLs87Jskdknw7Sbr7\n00n2m6Xzqtoryf2SvGpYryR3zWQK8yQ5PsmDhuVDhvUM2+827H9Ikjd19xXd/eUkZw31AAAAzN0s\noemH3X3pRm2zXp7390n+OslVw/oNk3y7u68c1s9LsuewvGeSc5Nk2H7psP+P25c5BgAAYK5mCU2f\nq6rfS7JDVe1fVS9J8u+rHVRV909yYXefdnWLnEVVHVFVp1bVqRdddNFaPCQAALAdmCU0PTqT+4mu\nSPJPmYwAPWaG4w5O8sCqOifJmzK5LO9FSXYZvuspSfZKcv6wfH6SvZMffxfUDZJ8a7p9mWN+rLuP\n7e4DuvuAPfbYY4byAAAAVjcamoYJF57e3U/s7tsPP0/q7u+v1nF3P7679+rufTOZyOFD3f2IJB9O\n8pBht0OTvHNY3jCsZ9j+oWGWvg1JHjbMrrdfkv2TfHLTniYAAMDm2XFsY3f/qKrutIUf83FJ3lRV\nz0xyepLjhvbjkry+qs5KcnEmQSvd/bmqOiHJ55NcmeTI7v7RFq4JAABgWaOhaXD6MMX4Pyf57lJj\nd79t1gfp7pOTnDwsn51lZr8bRq9+Z4Xjn5XkWbM+HgAAwJYyS2i6Tib3Ft11qq2TzByaAAAAtlar\nhqbuPmwtCgEAAFhEq4amqnrxMs2XJjm1u9+5zDYAAIBtxixTjl8nyW2SnDn83CqTab8Pr6q/n2Nt\nAAAA626We5puleTgpRnrqurlSf41yZ2SnDHH2gAAANbdLCNNuya5/tT6Tkl2G0LUFXOpCgAAYEHM\nMtL0t0k+XVUnJ6kkv5nk2VW1U5IPzLE2AACAdTfL7HnHVdWJ+cl3Kz2hu782LP/V3CoDAABYAKte\nnldVleRuSW49zJa3Y1X9zJfTAgAAbItmuafpH5LcMcnDh/XLk7xsbhUBAAAskFnuaTqwu29XVacn\nSXdfUlXXmnNdAAAAC2GWkaYfVtUOSTpJqmqPJFfNtSoAAIAFMUtoenGStye5UVU9K8nHkjx7rlUB\nAAAsiFlmz3tDVZ2WyWQQleRB3f2FuVcGAACwAFYMTVW129TqhUneOL2tuy+eZ2EAAACLYGyk6bRM\n7mOqJPskuWRY3iXJV5PsN/fqAAAA1tmK9zR1937d/QtJPpDkAd29e3ffMMn9k7x/rQoEAABYT7NM\nBHFQd5+4tNLd703y6/MrCQAAYHHM8j1NX6uqJyX5x2H9EUm+Nr+SAAAAFscsI00PT7JHJtOOv21Y\nfvg8iwIAAFgUs0w5fnGSo9agFgAAgIUzy0gTAADAdktoAgAAGCE0AQAAjFg1NFXVL1XVB6vqs8P6\nrYbZ9AAAALZ5s4w0vTLJ45P8MEm6+zNJHjbPogAAABbFLKHpet39yY3arpxHMQAAAItmltD0zar6\nxSSdJFX1kCQXzLUqAACABbHq9zQlOTLJsUl+parOT/LlJL8/16oAAAAWxCxfbnt2krtX1U5JrtHd\nl8+/LAAAgMWwYmiqqr9YoT1J0t0vmFNNAAAAC2NspGnn4fcvJ7l9kg3D+gOSbDwxBAAAwDZpxdDU\n3U9Lkqr6aJLbLV2WV1VPTfKeNakOAABgnc0ye96Nk/xgav0HQxsAAMA2b5bZ816X5JNV9fYkleSQ\nJK+dZ1EAAACLYpbZ855VVe9N8huZfFfTYd19+twrAwAAWACzjDQlyY+SXJVJaLpqfuUAAAAsllXv\naaqqo5K8IcnuSW6U5B+r6tHzLgwAAGARzDLSdHiSA7v7u0lSVc9N8h9JXjLPwgAAABbBLLPnVSaX\n5y350dAGAACwzZtlpOk1ST4xzJ6XJA9Kctz8SgIAAFgcs8ye94Kq+kiSg4cms+cBAADbjVlnz/t0\nkguW9q+qfbr7q3OrCgAAYEGsGpqGmfKOSfKN/OR+pk5yq/mWBgAAsP5mGWk6Kskvd/e35l0MAADA\nopll9rxzk1w670IAAAAW0SwjTWcnObmq3pPkiqXG7n7B3KoCAABYELOEpq8OP9cafgAAALYbs0w5\n/rS1KAQAAGARzXJPEwAAwHZLaAIAABghNAEAAIyY5ctt90jyJ0n2nd6/ux85v7IAAAAWwyyz570z\nyb8m+UCSH823HAAAgMUyS2i6Xnc/bu6VAAAALKBZ7ml6d1Xdd+6VAAAALKAVR5qq6vIknaSSPKGq\nrkjyw2G9u/vn1qZEAACA9bNiaOrundeyEAAAgEW06uV5VfXBWdoAAAC2RWOX510nyU5Jdq+qXTO5\nLC9Jfi7JnmtQGwAAwLobmz3vUUkek+SmST411X5ZkpfOsygAAIBFMXZP04uSvKiqHt3dL1nDmgAA\nABbG2OV5d+3uDyU5v6oevPH27n7bXCsDAABYAGOX5/1Wkg8lecAy2zqJ0AQAAGzzxi7PO2b4fdja\nlQMAALBYZply/EtV9Yaq+tOquuWsHVfVdarqk1X1n1X1uap62tC+X1V9oqrOqqo3V9W1hvZrD+tn\nDdv3nerr8UP7F6vqXpv+NAEAADbPqqEpyS2SvCLJDZM8bwhRb5/huCuS3LW7b53kNknuXVUHJXlu\nkhd2982TXJLk8GH/w5NcMrS/cNgvVXWLJA9Lcssk907yD1W1w6xPEAAA4OqYJTT9KMkPh99XJblw\n+BnVE98ZVq85/HSSuyZ5y9B+fJIHDcuHDOsZtt+tqmpof1N3X9HdX05yVpI7zFA3AADA1TY2EcSS\ny5KckeQFSV7Z3d+atfNhROi0JDdP8rIkX0ry7e6+ctjlvPzki3L3THJuknT3lVV1aSajW3sm+fhU\nt9PHAAAAzNUsI00PT/LRJP9vkjdV1dOq6m6zdN7dP+ru2yTZK5PRoV/Z7EpXUVVHVNWpVXXqRRdd\nNK+HAQAAtjOrhqbufmd3/1WSRyU5MckfJXn3pjxId387yYeT3DHJLlW1NMK1V5Lzh+Xzk+ydJMP2\nGyT51nT7MsdMP8ax3X1Adx+wxx57bEp5AAAAK5pl9ry3VtVZSV6U5HpJ/jDJrjMct0dV7TIsXzfJ\nPZJ8IZPw9JBht0OTvHNY3jCsZ9j+oe7uof1hw+x6+yXZP8knZ3t6AAAAV88s9zT9TZLTu/tHm9j3\nTZIcP9zXdI0kJ3T3u6vq85lc5vfMJKcnOW7Y/7gkrx8C2sWZzJiX7v5cVZ2Q5PNJrkxy5GbUAgAA\nsFlWDU3dfermdNzdn0ly22Xaz84ys9919/eT/M4KfT0rybM2pw4AAICrY5aJIAAAALZbQhMAAMCI\nWSaCOLiqdhqWf7+qXlBVN5t/aQAAAOtvlpGmlyf5XlXdOsljM/mC2tfNtSoAAIAFMUtounKY+vuQ\nJC/t7pcl2Xm+ZQEAACyGWaYcv7yqHp/kD5L8RlVdI8k151sWAADAYphlpOl3k1yR5JHd/fUkeyV5\n3lyrAgAAWBCrhqYhKL01ybWHpm8mefs8iwIAAFgUs8ye9ydJ3pLkFUPTnkneMc+iAAAAFsUsl+cd\nmeTgJJclSXefmeRG8ywKAABgUcwSmq7o7h8srVTVjkl6fiUBAAAsjllC00eq6glJrltV90jyz0ne\nNd+yAAAAFsMsoenoJBclOSPJo5KcmORJ8ywKAABgUaz6PU3dfVWSVw4/AAAA25VVQ1NVHZzkqUlu\nNuxfSbq7f2G+pQEAAKy/VUNTkuOS/HmS05L8aL7lAAAALJZZQtOl3f3euVcCAACwgGYJTR+uqucl\neVuSK5Yau/tTc6sKAABgQcwSmg4cfh8w1dZJ7rrlywEAAFgss8yed5e1KAQAAGARrfo9TVV146o6\nrqreO6zfoqoOn39pAAAA62+WL7d9bZL3JbnpsP7fSR4zr4IAAAAWySyhaffuPiHJVUnS3VfG1OMA\nAMB2YpbQ9N2qumEmkz+kqg5KculcqwIAAFgQs8ye9xdJNiT5xar6tyR7JHnIXKsCAABYELPMnvep\nqvqtJL+cpJJ8sbt/OPfKAAAAFsCqoamqHrxR0y9V1aVJzujuC+dTFgAAwGKY5fK8w5PcMcmHh/U7\nJzktyX5V9fTufv2cagMAAFh3s4SmHZP8and/I5l8b1OS1yU5MMlHkwhNAADANmuW2fP2XgpMgwuH\ntouTuLcJAADYps0y0nRyVb07yT8P6789tO2U5NtzqwwAAGABzBKajkzy4CR3GtZfl+St3d1J7jKv\nwgAAABbBLFOOd5K3Dj8AAADblVnuaQIAANhuCU0AAAAjVgxNVfXB4fdz164cAACAxTJ2T9NNqurX\nkzywqt6UpKY3dven5loZAADAAhgLTU9J8uQkeyV5wUbbOsld51UUAADAolgxNHX3W5K8paqe3N3P\nWMOaAAAAFsYsU44/o6oemOQ3h6aTu/vd8y0LAABgMaw6e15V/U2So5J8fvg5qqqePe/CAAAAFsGq\nI01J7pfkNt19VZJU1fFJTk/yhHkWBgAAsAhm/Z6mXaaWbzCPQgAAABbRLCNNf5Pk9Kr6cCbTjv9m\nkqPnWhUAAMCCmGUiiDdW1clJbj80Pa67vz7XqgAAABbELCNN6e4LkmyYcy0AAAALZ9Z7mgAAALZL\nQhMAAMCI0dBUVTtU1X+tVTEAAACLZjQ0dfePknyxqvZZo3oAAAAWyiwTQeya5HNV9ckk311q7O4H\nzq0qAACABTFLaHry3KsAAABYULN8T9NHqupmSfbv7g9U1fWS7DD/0gAAANbfqrPnVdWfJHlLklcM\nTXsmecc8iwIAAFgUs0w5fmSSg5NcliTdfWaSG82zKAAAgEUxS2i6ort/sLRSVTsm6fmVBAAAsDhm\nCU0fqaonJLluVd0jyT8nedd8ywIAAFgMs4Smo5NclOSMJI9KcmKSJ82zKAAAgEUxy+x5V1XV8Uk+\nkclleV/sbpfnAQAA24VVQ1NV3S/J/03ypSSVZL+qelR3v3fexQEAAKy3Wb7c9vlJ7tLdZyVJVf1i\nkvckEZoAAIBt3iz3NF2+FJgGZye5fE71AAAALJQVR5qq6sHD4qlVdWKSEzK5p+l3kpyyBrUBAACs\nu7HL8x4wtfyNJL81LF+U5LpzqwgAAGCBrBiauvuwtSwEAABgEa16T1NV7VdVL6iqt1XVhqWfGY7b\nu6o+XFWfr6rPVdVRQ/tuVXVSVZ05/N51aK+qenFVnVVVn6mq2031deiw/5lVdejVecIAAACbYpbZ\n896R5Lgk70py1Sb0fWWSx3b3p6pq5ySnVdVJSf4oyQe7+zlVdXQmX577uCT3SbL/8HNgkpcnObCq\ndktyTJIDMrmn6rSq2tDdl2xCLQAAAJtlltD0/e5+8aZ23N0XJLlgWL68qr6QZM8khyS587Db8UlO\nziQ0HZLkdcMX5368qnapqpsM+57U3RcnyRC87p3kjZtaEwAAwKaaJTS9qKqOSfL+JFcsNXb3p2Z9\nkKraN8ltk3wiyY2HQJUkX09y42F5zyTnTh123tC2UvvGj3FEkiOSZJ999pm1NAAAgFGzhKb/leQP\nktw1P7kck7e0AAAQlElEQVQ8r4f1VVXV9ZO8Ncljuvuyqvrxtu7uqupNqngF3X1skmOT5IADDtgi\nfQIAAMwSmn4nyS909w82tfOqumYmgekN3f22ofkbVXWT7r5guPzuwqH9/CR7Tx2+19B2fn5yOd9S\n+8mbWgsAAMDmWHX2vCSfTbLLpnZckyGl45J8obtfMLVpQ5KlGfAOTfLOqfY/HGbROyjJpcNlfO9L\ncs+q2nWYae+eQxsAAMDczTLStEuS/6qqU/LT9zQ9cJXjDs7ksr4zqurTQ9sTkjwnyQlVdXiSryR5\n6LDtxCT3TXJWku8lOWx4nIur6hlJThn2e/rSpBAAAADzNktoOmZzOu7ujyWpFTbfbZn9O8mRK/T1\n6iSv3pw6AAAAro5VQ1N3f2QtCgEAAFhEq4amqro8k9nykuRaSa6Z5Lvd/XPzLAwAAGARzDLStPPS\n8jC5wyFJDppnUQAAAItiltnzfqwn3pHkXnOqBwAAYKHMcnneg6dWr5HkgCTfn1tFAAAAC2SW2fMe\nMLV8ZZJzMrlEDwAAYJs3yz1Nh61FIQAAAItoxdBUVU8ZOa67+xlzqAcAAGChjI00fXeZtp2SHJ7k\nhkmEpi1g36Pfs+K2c55zvzWsBAAAWM6Koam7n7+0XFU7JzkqyWFJ3pTk+SsdBwAAsC0ZvaepqnZL\n8hdJHpHk+CS36+5L1qIwAACARTB2T9Pzkjw4ybFJ/ld3f2fNqgIAAFgQY19u+9gkN03ypCRfq6rL\nhp/Lq+qytSkPAABgfY3d0zQWqAAAALYLghEAAMAIoQkAAGCE0AQAADBCaAIAABghNAEAAIwQmgAA\nAEYITQAAACOEJgAAgBFCEwAAwAihCQAAYITQBAAAMEJoAgAAGCE0AQAAjBCaAAAARghNAAAAI4Qm\nAACAEUITAADACKEJAABghNAEAAAwQmgCAAAYITQBAACMEJoAAABGCE0AAAAjhCYAAIARQhMAAMAI\noQkAAGCE0AQAADBCaAIAABghNAEAAIwQmgAAAEYITQAAACOEJgAAgBFCEwAAwAihCQAAYITQBAAA\nMEJoAgAAGCE0AQAAjBCaAAAARghNAAAAI4QmAACAEUITAADACKEJAABghNAEAAAwQmgCAAAYITQB\nAACMEJoAAABGCE0AAAAjhCYAAIARQhMAAMAIoQkAAGDE3EJTVb26qi6sqs9Ote1WVSdV1ZnD712H\n9qqqF1fVWVX1maq63dQxhw77n1lVh86rXgAAgOXsOMe+X5vkpUleN9V2dJIPdvdzquroYf1xSe6T\nZP/h58AkL09yYFXtluSYJAck6SSnVdWG7r5kjnUvjH2Pfs+K2855zv3WsBIAANh+zW2kqbs/muTi\njZoPSXL8sHx8kgdNtb+uJz6eZJequkmSeyU5qbsvHoLSSUnuPa+aAQAANrbW9zTduLsvGJa/nuTG\nw/KeSc6d2u+8oW2ldgAAgDWxbhNBdHdncsndFlFVR1TVqVV16kUXXbSlugUAALZzax2avjFcdpfh\n94VD+/lJ9p7ab6+hbaX2n9Hdx3b3Ad19wB577LHFCwcAALZPax2aNiRZmgHv0CTvnGr/w2EWvYOS\nXDpcxve+JPesql2HmfbuObQBAACsibnNnldVb0xy5yS7V9V5mcyC95wkJ1TV4Um+kuShw+4nJrlv\nkrOSfC/JYUnS3RdX1TOSnDLs9/Tu3nhyCQAAgLmZW2jq7oevsOluy+zbSY5coZ9XJ3n1FiwNAABg\nZus2EQQAAMDWQGgCAAAYITQBAACMEJoAAABGCE0AAAAjhCYAAIARQhMAAMAIoQkAAGCE0AQAADBC\naAIAABghNAEAAIwQmgAAAEYITQAAACOEJgAAgBFCEwAAwAihCQAAYITQBAAAMEJoAgAAGCE0AQAA\njBCaAAAARuy43gWwefY9+j2j2895zv3WqBIAANi2GWkCAAAYITQBAACMEJoAAABGCE0AAAAjhCYA\nAIARQhMAAMAIoQkAAGCE0AQAADBCaAIAABghNAEAAIwQmgAAAEYITQAAACOEJgAAgBFCEwAAwAih\nCQAAYMSO610A87Hv0e9Zcds5z7nfGlYCAABbNyNNAAAAI4QmAACAEUITAADACKEJAABghNAEAAAw\nQmgCAAAYITQBAACMEJoAAABGCE0AAAAjhCYAAIARQhMAAMAIoQkAAGCE0AQAADBix/UugLW379Hv\nWXHbOc+53xpWAgAAi89IEwAAwAihCQAAYITQBAAAMEJoAgAAGGEiCH6KSSIAAOCnGWkCAAAYITQB\nAACMcHkeM3PpHgAA2yMjTQAAACOMNLFFGIUCAGBbZaQJAABghJEm5m5sFCoxEgUAwGITmlh3Lu0D\nAGCRuTwPAABgxFYz0lRV907yoiQ7JHlVdz9nnUtiDRiFAgBgvW0VoamqdkjysiT3SHJeklOqakN3\nf359K2M9rXav1OZYLYgJcQAA25+tIjQluUOSs7r77CSpqjclOSSJ0MQWdXWC2OYeK2wBACy2rSU0\n7Znk3Kn185IcuE61wBY1jxGzedmaAp5RQQBgS9laQtOqquqIJEcMq9+pqi+uZz1Tdk/yzfUugs3m\n/E2p5653BZts2fO3FT6P7ZH33tbN+du6OX9bL+du091slp22ltB0fpK9p9b3Gtp+rLuPTXLsWhY1\ni6o6tbsPWO862DzO39bN+dt6OXdbN+dv6+b8bb2cu/nZWqYcPyXJ/lW1X1VdK8nDkmxY55oAAIDt\nwFYx0tTdV1bV/5fkfZlMOf7q7v7cOpcFAABsB7aK0JQk3X1ikhPXu47NsHCXDLJJnL+tm/O39XLu\ntm7O39bN+dt6OXdzUt293jUAAAAsrK3lniYAAIB1ITTNUVXdu6q+WFVnVdXR610PP62q9q6qD1fV\n56vqc1V11NC+W1WdVFVnDr93Hdqrql48nM/PVNXt1vcZkCRVtUNVnV5V7x7W96uqTwzn6c3D5DGp\nqmsP62cN2/ddz7pJqmqXqnpLVf1XVX2hqu7o/bd1qKo/H/67+dmqemNVXcd7b3FV1aur6sKq+uxU\n2ya/16rq0GH/M6vq0PV4LtujFc7f84b/dn6mqt5eVbtMbXv8cP6+WFX3mmr3ufRqEJrmpKp2SPKy\nJPdJcoskD6+qW6xvVWzkyiSP7e5bJDkoyZHDOTo6yQe7e/8kHxzWk8m53H/4OSLJy9e+ZJZxVJIv\nTK0/N8kLu/vmSS5JcvjQfniSS4b2Fw77sb5elORfuvtXktw6k/Po/bfgqmrPJH+W5IDu/rVMJmh6\nWLz3Ftlrk9x7o7ZNeq9V1W5JjklyYJI7JDlmKWgxd6/Nz56/k5L8WnffKsl/J3l8kgyfYx6W5JbD\nMf8w/OOiz6VXk9A0P3dIclZ3n93dP0jypiSHrHNNTOnuC7r7U8Py5Zl8YNszk/N0/LDb8UkeNCwf\nkuR1PfHxJLtU1U3WuGymVNVeSe6X5FXDeiW5a5K3DLtsfP6Wzutbktxt2J91UFU3SPKbSY5Lku7+\nQXd/O95/W4sdk1y3qnZMcr0kF8R7b2F190eTXLxR86a+1+6V5KTuvri7L8nkQ/vGH+SZg+XOX3e/\nv7uvHFY/nsl3mCaT8/em7r6iu7+c5KxMPpP6XHo1CU3zs2eSc6fWzxvaWEDD5SK3TfKJJDfu7guG\nTV9PcuNh2TldPH+f5K+TXDWs3zDJt6f+RzJ9jn58/obtlw77sz72S3JRktcMl1e+qqp2ivffwuvu\n85P8XZKvZhKWLk1yWrz3tjab+l7zHlxcj0zy3mHZ+ZsToYntXlVdP8lbkzymuy+b3taT6SVNMbmA\nqur+SS7s7tPWuxY2y45Jbpfk5d192yTfzU8uD0ri/beohkuyDskk+N40yU4x4rBV817belXVEzO5\n3eAN613Ltk5omp/zk+w9tb7X0MYCqaprZhKY3tDdbxuav7F02c/w+8Kh3TldLAcneWBVnZPJZQZ3\nzeQemV2GS4aSnz5HPz5/w/YbJPnWWhbMTzkvyXnd/Ylh/S2ZhCjvv8V39yRf7u6LuvuHSd6WyfvR\ne2/rsqnvNe/BBVNVf5Tk/kke0T/5DiHnb06Epvk5Jcn+w2xC18rkprwN61wTU4Zr6o9L8oXufsHU\npg1JlmYFOjTJO6fa/3CYWeigJJdOXdrAGuvux3f3Xt29bybvrw919yOSfDjJQ4bdNj5/S+f1IcP+\n/mV1nXT315OcW1W/PDTdLcnn4/23NfhqkoOq6nrDf0eXzp333tZlU99r70tyz6radRhtvOfQxjqo\nqntncnn6A7v7e1ObNiR52DBr5X6ZTOjxyfhcerX5cts5qqr7ZnLPxQ5JXt3dz1rnkphSVXdK8q9J\nzshP7ol5Qib3NZ2QZJ8kX0ny0O6+ePhw8NJMLkP5XpLDuvvUNS+cn1FVd07yl919/6r6hUxGnnZL\ncnqS3+/uK6rqOklen8m9axcneVh3n71eNZNU1W0ymcTjWknOTnJYJv+Y5/234KrqaUl+N5PLgk5P\n8seZ3B/hvbeAquqNSe6cZPck38hkFrx3ZBPfa1X1yEz+P5kkz+ru16zl89herXD+Hp/k2vnJqO3H\nu/tPh/2fmMl9TldmcuvBe4d2n0uvBqEJAABghMvzAAAARghNAAAAI4QmAACAEUITAADACKEJAABg\nhNAEwJqoqq6q50+t/2VVPXUT+/jOJu5/TlWdMfx8vqqeOUyBDQAzE5oAWCtXJHlwVe2+xo97l+7+\nX0nukOQXkrxiS3RaVTtuiX4AWHxCEwBr5cokxyb58403VNW+VfWhqvpMVX2wqvYZ2verqv8YRoqe\nudExf1VVpwzHPG21B+/u7yT50yQPqqrdxvqoqidX1Rer6mNV9caq+suh/eSq+vuqOjXJUVW1R1W9\ndejjlKo6eNhvp6p6dVV9sqpOr6pDNvuvBsC6E5oAWEsvS/KIqrrBRu0vSXJ8d98qyRuSvHhof1GS\nlw8jRRcs7VxV90yyfyajR7dJ8r+r6jdXe/DuvizJl5Psv1IfVXX7JL+d5NZJ7pPkgI26uVZ3H9Dd\nzx/qe2F3Lx3zqmGfJyb5UHffIcldkjyvqnZarT4AFpNLCwBYM919WVW9LsmfJfmfqU13TPLgYfn1\nSf52WD44kzCy1P7cYfmew8/pw/r1MwlAH52hjFqlj52TvLO7v5/k+1X1ro2Of/PU8t2T3KJqqcv8\nXFVdf+j3gUsjVEmuk2SfJF+YoT4AFozQBMBa+/skn0rymhn372XaKsnfdPcm3Z9UVTsn2TfJf6/U\nR1U9ZpVuvju1fI0kBw0Ba7qPSvLb3f3FTakPgMXk8jwA1lR3X5zkhCSHTzX/e5KHDcuPSPKvw/K/\nbdS+5H1JHjmM6qSq9qyqG4097rDvPyR5R3dfMtLHvyV5QFVdZ9h2/5Fu35/k0VOPcZup+h49hKdU\n1W3HagNgsQlNAKyH5yeZnkXv0UkOq6rPJPmDJEcN7UclObKqzkiy59LO3f3+JP+U5D+GbW/J5LK6\n5Xy4qj6b5JNJvprkUWN9dPcpSTYk+UyS9yY5I8mlK/T9Z0kOGCaS+HwmE00kyTOSXDPJZ6rqc8M6\nAFup6l7uqgcA2H5V1fW7+ztVdb1M7pM6ors/td51AbA+3NMEAD/r2Kq6RSYTOBwvMAFs34w0AQAA\njHBPEwAAwAihCQAAYITQBAAAMEJoAgAAGCE0AQAAjBCaAAAARvz/kSYw4/Pwl98AAAAASUVORK5C\nYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f2788f9aba8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plot node degree to check if distribution follows power law\n",
    "node_degrees = ppi_network.sum()\n",
    "fig = plt.figure(figsize=(14, 8))\n",
    "bins = np.linspace(1,node_degrees.max(), 100)\n",
    "ax = plt.hist(node_degrees, bins)\n",
    "plt.xlabel('Node Degree')\n",
    "plt.ylabel('Number of genes with node degree')\n",
    "plt.title('Node Degree Distribution of Genes in PPI')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Remove gene expression for non-nodes\n",
    "Genes that have no interactions reported are not relevant for any network algorithm. I can kick out those genes for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Discard gene expression for 19248 genes because they are not in network\n",
      "Left with 10823 genes for which we have GE and interactions\n"
     ]
    }
   ],
   "source": [
    "# remove all genes that are not present in network\n",
    "gene_expression_short = ge_final[ge_final.index.isin(ppi_network.index)]\n",
    "print (\"Discard gene expression for {} genes because they are not in network\".\\\n",
    "       format(ge_final.shape[0]-gene_expression_short.shape[0])\n",
    "      )\n",
    "print (\"Left with {} genes for which we have GE and interactions\".format(gene_expression_short.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Split Replicates for Agreement Test\n",
    "The features (gene expression) comes from a dual RNA-seq experiment with three replicates.\n",
    "In order to evaluate the GCN performance, it is helpful to see how consistent the results are for each of the replicates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((10823, 8), (10823, 8), (10823, 8))"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select columns for replicates 2 and 3 easily by suffix .1/.2\n",
    "rep2 = gene_expression_short.filter(regex=\"\\.1$\", axis=1)\n",
    "rep3 = gene_expression_short.filter(regex=\"\\.2$\", axis=1)\n",
    "# select columns for replicate 1 by using all left over cols\n",
    "cond = ~(gene_expression_short.columns.isin(rep2) | gene_expression_short.columns.isin(rep3))\n",
    "rep1 = gene_expression_short.loc[:,cond].drop('Name', axis=1) # drop gene name\n",
    "rep1.shape, rep2.shape, rep3.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Get Negatives for Training\n",
    "I want to maybe use the low-variance genes as negatives for a training and testing set. Or, I could use the genes that are expressed in GFP-minus but not in GFP-plus instead."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(141, 25)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "low_var_genes = gene_expression_short[gene_expression_short.var(axis=1) < .5]\n",
    "low_var_genes.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(396, 25)\n",
      "In the list, there are 33 duplicates!\n",
      "After merge with IDs, there are 33 duplicates!\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# read negatives and add some information (names and numbers)\n",
    "negative_gene_names = pd.DataFrame.from_csv('../data/labels/non_involved_gfpm_ctrl.txt', header=0)\n",
    "negative_gene_names['index'] = np.arange(0, negative_gene_names.shape[0])\n",
    "negative_gene_names['gene_name'] = negative_gene_names.index\n",
    "negative_gene_names.set_index('index', inplace=True)\n",
    "\n",
    "# merge with gene expression to find the genes we actually have in the network\n",
    "negatives_all = gene_expression[gene_expression.Name.isin(negative_gene_names.gene_name)]\n",
    "print (negatives_all.shape)\n",
    "\n",
    "no_duplicates_names = negatives_all.Name.duplicated().sum()\n",
    "print (\"In the list, there are {} duplicates!\".format(no_duplicates_names))\n",
    "no_duplicates_merge = negatives_all.Name.duplicated().sum()\n",
    "print (\"After merge with IDs, there are {} duplicates!\".format(no_duplicates_merge))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-366...done.\n",
      "Finished.\n",
      "33 input query terms found dup hits:\n",
      "\t[('SNORA20', 4), ('SCARNA20', 7), ('FTH1P10', 2), ('SCARNA11', 4), ('EPS15P1', 2), ('RNU6-15P', 2), \n",
      "72 input query terms found no hit:\n",
      "\t['FAM65B', 'AC022816.2', 'RP4-799D16.1', 'SNORA34', 'AC007099.1', 'CH17-76K2.6', 'RP11-292D4.3', 'RP\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "Converted 366 gene names to IDs and received 341 Ensembl IDs\n",
      "Of these 341 IDs, I define 141 negatives that have gene expression\n"
     ]
    }
   ],
   "source": [
    "# convert the list of names to proper Ensembl IDs\n",
    "mg = mygene.MyGeneInfo()\n",
    "list_of_neg_genenames = negative_gene_names.gene_name.tolist()\n",
    "f = lambda x: x['ensembl'][0]['gene'] if type(x['ensembl']) is list else x['ensembl']['gene']\n",
    "res = mg.querymany(list_of_neg_genenames, scopes='symbol', fields='ensembl.gene', species='human')\n",
    "list_of_neg_ids = [f(x) for x in res if 'ensembl' in x]\n",
    "print (\"Converted {} gene names to IDs and received {} Ensembl IDs\".format(len(list_of_neg_genenames),\n",
    "                                                                           len(list_of_neg_ids))\n",
    "      )\n",
    "\n",
    "# finally, define the negatives\n",
    "if NEGATIVE_SET_LOWVAR:\n",
    "    negatives = gene_expression_short[gene_expression_short.index.isin(low_var_genes.index)]\n",
    "else:\n",
    "    negatives = gene_expression_short[gene_expression_short.index.isin(negatives_all.index)]\n",
    "print (\"Of these {} IDs, I define {} negatives that have gene expression\".format(len(list_of_neg_ids),\n",
    "                                                                                 negatives.shape[0])\n",
    "      )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Load the Labels and Do Train Test Split\n",
    "Finally, I want to load the labels that I obtained by literature search and that I got from Wigo.\n",
    "Once loaded and joined, I derive one big dataframe with a label column to make sure they are in the same order as the rest of the data, especially the PPI network.\n",
    "Then, I calculate train and test splits (**but I have to be careful because of the extremely high class imbalance**).\n",
    "And last but not least, I have to build the training and testing masks that are needed by the GCN algorithm."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def build_mask(names_labeled, X):\n",
    "    mask = names_labeled[names_labeled.isin(X)].label\n",
    "    # NaN means not being in the set\n",
    "    mask[~mask.isnull()] = 1\n",
    "    mask[mask.isnull()] = 0\n",
    "    # sanity check\n",
    "    assert (np.all(mask.index == names_labeled.index))\n",
    "    return mask.values\n",
    "\n",
    "def build_y(names_labeled, X):\n",
    "    y = names_labeled[names_labeled.isin(X)].infection_gene\n",
    "    # NaN means not being in the set\n",
    "    y[y.isnull()] = 0\n",
    "    # sanity check\n",
    "    assert (np.all(y.index == names_labeled.index))\n",
    "    y_onehot = pd.get_dummies(y).values\n",
    "    return y_onehot[:, [1,0]] # switch the positive and negative columns\n",
    "\n",
    "def balanced_stratified_split(X, test_ratio):\n",
    "    # calculate number of labels for both sets\n",
    "    num_labels_total = X.label.sum()\n",
    "    num_labels_train = int(num_labels_total * (1-test_ratio))\n",
    "    num_labels_test = num_labels_total - num_labels_train\n",
    "    num_infection_genes = X.infection_gene.sum()\n",
    "    num_infection_train = int(num_infection_genes*(1-test_ratio))\n",
    "    num_infection_test = num_infection_genes - num_infection_train\n",
    "    print (num_labels_total, num_labels_train, num_labels_test)\n",
    "    print (num_infection_genes, num_infection_train, num_infection_test)\n",
    "    # select random training set (size 2*num_labels_train)\n",
    "    X_train = X[X.infection_gene].sample(n=num_infection_train)\n",
    "    print (X_train.shape)\n",
    "    X_train = X_train.append(X[X.label & ~X.infection_gene].sample(n=num_infection_train))\n",
    "    print (X_train.shape)\n",
    "    # rest is test\n",
    "    X_test = X[~X.index.isin(X_train.index)]\n",
    "    return X_train, X_test\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collected 46 genes involved in legionella infection\n",
      "Out of the 46 infection genes, 38 are in the network\n",
      "Split Training and Testing with 40.0% test nodes\n",
      "Training Nodes: 107\t#Infection Genes in Train Set: 23\n",
      "Testing Nodes: 72\t#Infection Genes in Test Set: 15\n"
     ]
    }
   ],
   "source": [
    "# load the labels from Wigos list and my literature search (together with Wigos extension)\n",
    "infection_genes = pd.DataFrame.from_csv('../data/labels/labels_wigo_codingonly.csv', sep='\\t', index_col=3)\n",
    "my_infection_genes = pd.DataFrame.from_csv('../data/labels/labels_wigo_extended.tsv',\n",
    "                                           sep='\\t',\n",
    "                                           index_col=1\n",
    "                                          )\n",
    "infection_genes = infection_genes.append(my_infection_genes)\n",
    "print (\"Collected {} genes involved in legionella infection\".format(infection_genes.shape[0]))\n",
    "\n",
    "# load the gene names and add label column to it\n",
    "names_labeled = pd.DataFrame(gene_expression_short.Name)\n",
    "names_labeled['infection_gene'] = names_labeled.index.isin(infection_genes.index) # positive for infection genes\n",
    "names_labeled['label'] = False\n",
    "names_labeled.loc[names_labeled.index.isin(negatives.index), 'label'] = True # negatives are only labels\n",
    "names_labeled.label = names_labeled.label | names_labeled.infection_gene # both infection and negatives are labels\n",
    "\n",
    "print (\"Out of the {} infection genes, {} are in the network\".format(infection_genes.shape[0],\n",
    "                                                                     names_labeled.infection_gene.sum())\n",
    "      )\n",
    "\n",
    "# split for training, testing & validation\n",
    "if BALANCE:\n",
    "    X_train, X_test = balanced_stratified_split(names_labeled, TEST_RATIO)\n",
    "else:\n",
    "    X_train, X_test = train_test_split(names_labeled[names_labeled.label],\n",
    "                                       test_size=TEST_RATIO,\n",
    "                                       stratify=names_labeled[names_labeled.label].infection_gene\n",
    "                                      )\n",
    "if VAL_SIZE > 0:\n",
    "    X_val = X_train[-VAL_SIZE:]\n",
    "    X_train = X_train[:-VAL_SIZE]\n",
    "else:\n",
    "    X_val = None\n",
    "print (\"Split Training and Testing with {}% test nodes\".format(TEST_RATIO*100.))\n",
    "print (\"Training Nodes: {}\\t#Infection Genes in Train Set: {}\".format(X_train.shape[0],\n",
    "                                                                 X_train.infection_gene.sum())\n",
    "      )\n",
    "print (\"Testing Nodes: {}\\t#Infection Genes in Test Set: {}\".format(X_test.shape[0],\n",
    "                                                               X_test.infection_gene.sum())\n",
    "      )\n",
    "\n",
    "# construct masks\n",
    "train_mask = build_mask(names_labeled, X_train)\n",
    "test_mask = build_mask(names_labeled, X_test)\n",
    "if X_val is not None:\n",
    "    val_mask = build_mask(features_labeled, X_val)\n",
    "\n",
    "# construct labels\n",
    "y_train = build_y(names_labeled, X_train)\n",
    "y_test = build_y(names_labeled, X_test)\n",
    "if X_val is not None:\n",
    "    y_val = pd.get_dummies(features_labeled[features_labeled.isin(X_val)].label).values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Write to hdf5 File on Disk\n",
    "The last thing to do is to write the data to disk, using the fast and efficient hdf5 format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Pam3T16</th>\n",
       "      <th>Pam3T8</th>\n",
       "      <th>Pam3T16.1</th>\n",
       "      <th>Pam3T8.1</th>\n",
       "      <th>Pam3T16.2</th>\n",
       "      <th>Pam3T8.2</th>\n",
       "      <th>ControlT8</th>\n",
       "      <th>ControlT16</th>\n",
       "      <th>ControlT8.1</th>\n",
       "      <th>...</th>\n",
       "      <th>gfpmT8.1</th>\n",
       "      <th>gfpmT16.1</th>\n",
       "      <th>gfpmT8.2</th>\n",
       "      <th>gfpmT16.2</th>\n",
       "      <th>gfppT8</th>\n",
       "      <th>gfppT16</th>\n",
       "      <th>gfppT8.1</th>\n",
       "      <th>gfppT16.1</th>\n",
       "      <th>gfppT8.2</th>\n",
       "      <th>gfppT16.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ensembl-ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>6.736327</td>\n",
       "      <td>8.519666</td>\n",
       "      <td>10.773706</td>\n",
       "      <td>10.813725</td>\n",
       "      <td>2.427830</td>\n",
       "      <td>7.425264</td>\n",
       "      <td>9.417385</td>\n",
       "      <td>7.920964</td>\n",
       "      <td>10.812953</td>\n",
       "      <td>...</td>\n",
       "      <td>12.529836</td>\n",
       "      <td>6.769370</td>\n",
       "      <td>13.440488</td>\n",
       "      <td>5.248289</td>\n",
       "      <td>14.583375</td>\n",
       "      <td>13.538040</td>\n",
       "      <td>18.220493</td>\n",
       "      <td>14.908853</td>\n",
       "      <td>16.156797</td>\n",
       "      <td>30.745221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>97.676739</td>\n",
       "      <td>73.025712</td>\n",
       "      <td>132.362676</td>\n",
       "      <td>110.840686</td>\n",
       "      <td>87.401897</td>\n",
       "      <td>87.865624</td>\n",
       "      <td>92.725018</td>\n",
       "      <td>105.132797</td>\n",
       "      <td>121.218890</td>\n",
       "      <td>...</td>\n",
       "      <td>127.983328</td>\n",
       "      <td>147.421826</td>\n",
       "      <td>104.163784</td>\n",
       "      <td>86.596770</td>\n",
       "      <td>298.959192</td>\n",
       "      <td>239.172041</td>\n",
       "      <td>268.101547</td>\n",
       "      <td>155.611158</td>\n",
       "      <td>283.898001</td>\n",
       "      <td>247.298513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>84.204085</td>\n",
       "      <td>74.242807</td>\n",
       "      <td>61.564035</td>\n",
       "      <td>68.486928</td>\n",
       "      <td>78.499852</td>\n",
       "      <td>92.815800</td>\n",
       "      <td>86.929705</td>\n",
       "      <td>72.728852</td>\n",
       "      <td>66.585024</td>\n",
       "      <td>...</td>\n",
       "      <td>76.074006</td>\n",
       "      <td>63.180782</td>\n",
       "      <td>117.604272</td>\n",
       "      <td>94.469203</td>\n",
       "      <td>83.125239</td>\n",
       "      <td>79.724014</td>\n",
       "      <td>76.786365</td>\n",
       "      <td>62.430824</td>\n",
       "      <td>85.400212</td>\n",
       "      <td>93.572410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>FGR</td>\n",
       "      <td>53.890614</td>\n",
       "      <td>150.919804</td>\n",
       "      <td>143.136382</td>\n",
       "      <td>362.259802</td>\n",
       "      <td>156.999704</td>\n",
       "      <td>160.880721</td>\n",
       "      <td>62.299622</td>\n",
       "      <td>61.207450</td>\n",
       "      <td>189.511222</td>\n",
       "      <td>...</td>\n",
       "      <td>333.830638</td>\n",
       "      <td>257.988195</td>\n",
       "      <td>186.486774</td>\n",
       "      <td>70.851902</td>\n",
       "      <td>129.792040</td>\n",
       "      <td>97.774734</td>\n",
       "      <td>272.005939</td>\n",
       "      <td>245.996082</td>\n",
       "      <td>140.794944</td>\n",
       "      <td>101.592903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <td>CFH</td>\n",
       "      <td>18.524899</td>\n",
       "      <td>21.907714</td>\n",
       "      <td>43.094825</td>\n",
       "      <td>53.167483</td>\n",
       "      <td>29.943242</td>\n",
       "      <td>23.513336</td>\n",
       "      <td>4.346485</td>\n",
       "      <td>2.880351</td>\n",
       "      <td>15.365775</td>\n",
       "      <td>...</td>\n",
       "      <td>11.634848</td>\n",
       "      <td>17.299500</td>\n",
       "      <td>6.720244</td>\n",
       "      <td>6.997719</td>\n",
       "      <td>5.833350</td>\n",
       "      <td>7.521133</td>\n",
       "      <td>5.205855</td>\n",
       "      <td>11.181640</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.336749</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Name    Pam3T16      Pam3T8   Pam3T16.1    Pam3T8.1  \\\n",
       "Ensembl-ID                                                               \n",
       "ENSG00000000003  TSPAN6   6.736327    8.519666   10.773706   10.813725   \n",
       "ENSG00000000419    DPM1  97.676739   73.025712  132.362676  110.840686   \n",
       "ENSG00000000457   SCYL3  84.204085   74.242807   61.564035   68.486928   \n",
       "ENSG00000000938     FGR  53.890614  150.919804  143.136382  362.259802   \n",
       "ENSG00000000971     CFH  18.524899   21.907714   43.094825   53.167483   \n",
       "\n",
       "                  Pam3T16.2    Pam3T8.2  ControlT8  ControlT16  ControlT8.1  \\\n",
       "Ensembl-ID                                                                    \n",
       "ENSG00000000003    2.427830    7.425264   9.417385    7.920964    10.812953   \n",
       "ENSG00000000419   87.401897   87.865624  92.725018  105.132797   121.218890   \n",
       "ENSG00000000457   78.499852   92.815800  86.929705   72.728852    66.585024   \n",
       "ENSG00000000938  156.999704  160.880721  62.299622   61.207450   189.511222   \n",
       "ENSG00000000971   29.943242   23.513336   4.346485    2.880351    15.365775   \n",
       "\n",
       "                    ...        gfpmT8.1   gfpmT16.1    gfpmT8.2  gfpmT16.2  \\\n",
       "Ensembl-ID          ...                                                      \n",
       "ENSG00000000003     ...       12.529836    6.769370   13.440488   5.248289   \n",
       "ENSG00000000419     ...      127.983328  147.421826  104.163784  86.596770   \n",
       "ENSG00000000457     ...       76.074006   63.180782  117.604272  94.469203   \n",
       "ENSG00000000938     ...      333.830638  257.988195  186.486774  70.851902   \n",
       "ENSG00000000971     ...       11.634848   17.299500    6.720244   6.997719   \n",
       "\n",
       "                     gfppT8     gfppT16    gfppT8.1   gfppT16.1    gfppT8.2  \\\n",
       "Ensembl-ID                                                                    \n",
       "ENSG00000000003   14.583375   13.538040   18.220493   14.908853   16.156797   \n",
       "ENSG00000000419  298.959192  239.172041  268.101547  155.611158  283.898001   \n",
       "ENSG00000000457   83.125239   79.724014   76.786365   62.430824   85.400212   \n",
       "ENSG00000000938  129.792040   97.774734  272.005939  245.996082  140.794944   \n",
       "ENSG00000000971    5.833350    7.521133    5.205855   11.181640    0.000000   \n",
       "\n",
       "                  gfppT16.2  \n",
       "Ensembl-ID                   \n",
       "ENSG00000000003   30.745221  \n",
       "ENSG00000000419  247.298513  \n",
       "ENSG00000000457   93.572410  \n",
       "ENSG00000000938  101.592903  \n",
       "ENSG00000000971    1.336749  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_expression_short.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing container to ../data/preprocessing/legionella_neglowvar_unbalanced.h5\n"
     ]
    }
   ],
   "source": [
    "# remove second column of labels\n",
    "y_train_one = y_train[:, 0].reshape(-1, 1)\n",
    "y_test_one = y_test[:, 0].reshape(-1, 1)\n",
    "\n",
    "# create gene names\n",
    "gene_names_ids = np.vstack((gene_expression_short.index, gene_expression_short.Name)).transpose(1, 0)\n",
    "\n",
    "# remove name column from gene expression because otherwise hdf5 can't save it\n",
    "features = gene_expression_short.drop('Name', axis=1)\n",
    "string_dt = h5py.special_dtype(vlen=str)\n",
    "\n",
    "fname = '../data/preprocessing/legionella_{}_{}.h5'\n",
    "fname = fname.format('neglowvar' if NEGATIVE_SET_LOWVAR else 'negnopam', 'balanced' if BALANCE else 'unbalanced')\n",
    "print (\"Writing container to {}\".format(fname))\n",
    "f = h5py.File(fname, 'w')\n",
    "# add ppi network\n",
    "f.create_dataset('network', data=ppi_network, shape=ppi_network.shape)\n",
    "# add features and replicates\n",
    "f.create_dataset('features', data=features, shape=features.shape)\n",
    "f.create_dataset('features_rep1', data=rep1, shape=rep1.shape)\n",
    "f.create_dataset('features_rep2', data=rep2, shape=rep2.shape)\n",
    "f.create_dataset('features_rep3', data=rep3, shape=rep3.shape)\n",
    "# add gene names (same order as features & network)\n",
    "f.create_dataset('gene_names', data=gene_names_ids, dtype=string_dt)\n",
    "# add labels\n",
    "f.create_dataset('y_train', data=y_train_one, shape=y_train_one.shape)\n",
    "f.create_dataset('y_test', data=y_test_one, shape=y_test_one.shape)\n",
    "if X_val is not None:\n",
    "    f.create_dataset('y_val', data=y_val, shape=y_val.shape)\n",
    "\n",
    "f.create_dataset('mask_train', data=train_mask, shape=train_mask.shape)\n",
    "f.create_dataset('mask_test', data=test_mask, shape=test_mask.shape)\n",
    "if X_val is not None:\n",
    "    f.create_dataset('mask_val', data=val_mask, shape=val_mask.shape)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Not a file id (Not a file id)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-22-fa01cc0d8510>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/home/sasse/.local/lib/python3.6/site-packages/h5py/_hl/files.py\u001b[0m in \u001b[0;36mclose\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    285\u001b[0m             \u001b[0;31m# Close file-resident objects first, then the files.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    286\u001b[0m             \u001b[0;31m# Otherwise we get errors in MPI mode.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 287\u001b[0;31m             \u001b[0mid_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mh5f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_obj_ids\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m~\u001b[0m\u001b[0mh5f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mOBJ_FILE\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    288\u001b[0m             \u001b[0mfile_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mh5f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_obj_ids\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mh5f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mOBJ_FILE\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    289\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mh5py/_objects.pyx\u001b[0m in \u001b[0;36mh5py._objects.with_phil.wrapper (/tmp/pip-s_7obrrg-build/h5py/_objects.c:2840)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mh5py/_objects.pyx\u001b[0m in \u001b[0;36mh5py._objects.with_phil.wrapper (/tmp/pip-s_7obrrg-build/h5py/_objects.c:2798)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mh5py/h5f.pyx\u001b[0m in \u001b[0;36mh5py.h5f.get_obj_ids (/tmp/pip-s_7obrrg-build/h5py/h5f.c:3419)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Not a file id (Not a file id)"
     ]
    }
   ],
   "source": [
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10823, 1)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_train_one.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10823, 2)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_names_ids = np.vstack((gene_expression_short.index, gene_expression_short.Name)).transpose(1, 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([   22, 10801], dtype=uint64),\n",
       " array([   16, 10807], dtype=uint64),\n",
       " (10823,),\n",
       " (10823,))"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_train.sum(axis=0), y_test.sum(axis=0), train_mask.shape, test_mask.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
